---
title: "Polymethylation scores results"
author: "F blostein"
date: "9/30/2021"
output:
  html_document:
    code_folding: show
    highlight: tango
    number_sections: no
    theme: sandstone
  pdf_document: default
  word_document: default
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message=FALSE)

library(ggplot2)
library(kableExtra)
library(ggpubr)
library(tidyr)
library(stringr)
library(dplyr)
library(compareGroups)
library(kableExtra)
library(gtsummary)
library(gtools)
library(haven)
library(sjlabelled)
library(matchmaker)
library(EpiDISH)
library("PerformanceAnalytics")
library(DescTools)
library(purrr)
library(broom)
library(broom.mixed)
library(ggplot2)
library(multcomp)
library(ggpubr)
library(cowplot)
library(geepack)
library(lme4)
library(pROC)
library(lmerTest)
library(stringr)
library(sva)
library(gtsummary)
library(gt)
library(here)
library(DiagrammeRsvg)
library(xml2)
library(ggrepel)
```

```{r temp read n data}
basedir=here::here()
outdir=file.path(gsub('Code', 'Output', basedir), 'Figures')
datadir=gsub('Code', 'Data', basedir)
load(paste0(datadir, '/CreatedData/allPhenoData.Rdata'))
load(paste0(datadir, '/CreatedData/allmethyl.Rdata'))
load(paste0(datadir, '/CreatedData/completeCaseSVA.Rdata'))
load(paste0(datadir, '/CreatedData/methylation_summarymodels.Rdata'))
```

```{r plotting variables}
theme_set(theme_bw())
```

```{r}
modeldata=modeldata %>% mutate(childteen=factor(childteen, levels=c('Age 9', 'Age 15')))
n_total=modeldata %>% dplyr::select(id)%>%unique()%>%nrow()
n_obs=modeldata%>%nrow()

all_methyl=c("globalmethylation", "pediatric", "grim", "Epithelial.cells_saliva", "Leukocytes_saliva", "SSnewborn_center", "SSnewbornCT_center", "SSolder_center", "cg05575921")
main_methyl=c("globalmethylation", "pediatric", "SSnewbornCT_center", "cg05575921")


all_methyl_labels=modeldata%>%dplyr::select(any_of(all_methyl))%>%get_label()
all_methyl_labels[str_detect(all_methyl_labels, 'Polymethylation', )]=paste0(rep('Score:', 3), c('Sustained smoking \n newborns', 'Sustained smoking \n newborns (w/ CT control)', 'Sustained smoking \n older children'))
names(all_methyl_labels)=all_methyl

all_methyl_levels=c("SSnewbornCT_center", "SSnewborn_center",  "SSolder_center",
                                       "pediatric" , "grim", "cg05575921", 
                                      "globalmethylation","Epithelial.cells_saliva", "Leukocytes_saliva")

main_methyl_labels=modeldata%>%dplyr::select(any_of(main_methyl))%>%get_label()

my_round=function(x, d){
  y=c()
  d_orig=d
  for(i in 1:length(x)){
    d=d_orig
    while(round(x[i], d)==0){
    d=d+1
    }
    y[i]=round(x[i], d)
  }
  return(y)
}
```

```{r figure counter}
main_fig=1
sup_fig=1
```

    
# Target journal and deadline 

[Clinical epigenetics](https://www.biomedcentral.com/collections/epibio): Special issue on biomarkers; due date December 1st


 "Epigenetic biomarkers have significantly contributed to improved understanding of the origins and progression of disease.  Moreover, increasing evidence shows that epigenetic biomarkers have potential for personalized medicine. However, routine use of epigenetic biomarkers in in-vitro diagnostics (IVD) is lagging behind the discoveries of new biomarkers. In this thematic series of Clinical Epigenetics, we discuss current progress in development of epigenetic biomarkers for clinical use in common complex diseases of ageing and welcome both original research and reviews.
The sections of the collection include:

  - Types of epigenetic biomarkers and readiness for IVD use
  
    - DNA methylation as biomarker
    
    - Histone modification as biomarker
    
    - Upcoming field of circulating free nucleosomes as biomarker
    
    - Progress towards clinical use of biomarker
    
  - *Biomarkers of environmental exposures and causality of the biomarker*
  
  - Disease predisposition biomarkers
  
  - Biomarkers used for disease detection
  
  - Clinical disease prevention and management: predictive and prognostic
  
  - Monitoring of chronic disease
  
  - *Single versus multiple biomarkers and risk scores*
  
  - Major stages of clinical validation of the biomarker
  
  - Biomarker detection technologies suitable for IVD use and significance of bioinformatics
  
  - Ethics and regulatory aspects of IVD epigenetic biomarkers

# Figure 1 - Selection of samples from the Fragile Families and Child Wellbeing (FFCW) study into this analysis. N represents the number of individuals at each step in the selection procedure, M represents the number of samples, as individuals with repeated measures can have more than one sample. 

```{r, eval=T}
#function
N_M_count=function(filter_statement, statement, df){
  paste0('m=', df %>% filter(eval(parse(text = filter_statement)))%>%nrow(), 
         ' samples & ', 
         'n=', df %>% filter(eval(parse(text = filter_statement))) %>% dplyr::select(idnum) %>%unique()%>%nrow(), 
         ' individuals with', statement)
}
#Make Labels
n_fullFF<-paste0("Fragile Families cohort \n", "n=", FF_labeled %>% nrow())
n_methylQC<-paste0("Cohort with 450K methylation data \n", "n=", methylCohort$idnum%>%unique()%>%length(), " individuals with m=", methylCohort$MethID%>%length(), " samples")

n_QCloss<-paste0("Methylation data QC: ", 
                 N_M_count("probe_fail_10==T & childteen!='M'", " with >10% of probes failing dropped, \n ", pdqc_all),
                 N_M_count("sex_new!=predicted_sex & probe_fail_10==F & childteen!='M'", 
                           "  with discordant sex samples dropped, \n ", pdqc_all),
                 N_M_count("predicted_sex_outlier==T & sex_new==predicted_sex & probe_fail_10==F & childteen!='M'", 
                           ' with abnormal sex chromosome intensity dropped & \n ', pdqc_all),
                 length(reps_remove), " duplicate samples dropped")

n_methylFF<-paste0("Cohort with quality controlled 450K data \n","n=", myFF %>% dplyr::select(id) %>% unique() %>% nrow(), " individuals with m=", myFF%>%nrow(), " samples")


n_missingmaternalprenataldata=paste0('Missing maternal data: \n', 
                                     N_M_count("is.na(smkPreg_binary)", " with missing prenatal maternal smoking \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & is.na(m1g2_YesNoPreg)", " with missing prenatal alcohol \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & is.na(m1g3_YesNoPreg)", 
                                               " with missing prenatal other drug use \n ", myFF),
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & is.na(cm1inpov)", 
                                               " with missing maternal income to poverty ratio data \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) &
                                               PostnatalMaternalSmokingAny=='Missing'", " with missing postnatal maternal smoking data \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) &
                                               PostnatalMaternalSmokingAny!='Missing' & SmkAtVisitPastmonth=='Missing'", 
                                               " with missing past month maternal/primary care giver smoking packs/day", myFF))

maternaldata=myFF %>% filter(!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) & PostnatalMaternalSmokingAny!='Missing' & SmkAtVisitPastmonth!='Missing')  
n_withmaternaldata=paste0('m=', maternaldata %>% nrow(), ' samples & n=', maternaldata %>% dplyr::select(idnum)%>%unique()%>%nrow(), ' individuals with maternal covariate data')
  
n_missingchilddata=paste0('Missing child data: \n', 
                          N_M_count("is.na(ancestry) | ancestry=='Missing PC data'", 
                                    " with missing ancestry/genetic data  \n ", maternaldata), 
                          N_M_count("!is.na(ancestry) & ancestry!='Missing PC data' & is.na(cm1bsex)", 
                                    " with missing child sex data \n ", maternaldata), 
                          N_M_count("is.na(ancestry) | ancestry=='Missing PC data' & !is.na(cm1bsex) & is.na(ChildAgeComposite)", 
                                    " with missing child age at sample \n ", maternaldata))

childdata= maternaldata %>% filter(ancestry!='Missing PC data' & !is.na(cm1bsex) & !is.na(ChildAgeComposite))
n_withchilddata=paste0('m=', childdata %>% nrow(), ' samples & n=', childdata %>% dplyr::select(idnum)%>%unique()%>%nrow(), ' individuals with child covariate data')

n_Child<-paste0('n=', childdata %>%filter(childteen=='C')  %>% nrow(), " individuals with a 9 year visit")
n_ChildSmkexcluded<-paste0("Exclude children who smoke: \n",'n=', childdata%>%filter(childteen=='C' & k5f1l=='2 no')%>% nrow(), 
                           " individuals with a 9 year visit")
n_ChildSmkexclusions<-paste0('n=', childdata%>%filter(k5f1l!= "2 no" & childteen=='C')%>%nrow(), 
                             " individuals smoking \n or missing child smoking data \n at 9 year visit")
n_Teen<-paste0('n=', childdata %>%filter(childteen=='T')   %>% nrow(), " individuals with a 15 year visit")
n_TeenSmkexcluded<-paste0("Exclude children who smoke: \n",'n=', childdata %>%filter(childteen=='T' & k6d40=="2 No") %>% nrow(), 
                          " individuals with a 15 year visit")
n_TeenSmkexclusions<-paste0('n=', childdata %>% filter(k6d40!="2 No" & childteen=='T') %>% nrow,
                            " individuals smoking \n or missing focal child smoking data \n at 15 year visit") 
if(childdata %>% filter((childteen=='C' & k5f1l=='2 no') | (childteen=='T' & k6d40=='2 No'))%>%nrow()!=nrow(modeldata)){stop('filter N and completecase data set N do not match')}


```

```{r, fig.height=12}

graph1<-DiagrammeR::grViz("
   digraph graph2 {
   compound=true;
   graph [layout = dot]
   
   node [shape = rectangle, width=4]
   a [label = '@@1']
   b [label = '@@2']
   c [label = '@@3']
   d [label = '@@4']
   e [label = '@@5']
   f [label = '@@6']
   g [label = '@@7']
   h [label = '@@11']
   i [label = '@@12']
   

   #fake nodes for connecting to edges
   node [shape=point, width=0.01, height=0.01]
   a1
   b1
   c1
   d1
   f1 [group=g1]
   g1
   h1
  
  #exclusion nodes
  node [shape = rectangle, width=4]
  aa [label = '@@8']
  bb [label = '@@9']
  cc [label = '@@10']
  hh [label = '@@13']
  ii [label = '@@14']
    
  #draw connections
#  a->m
#  m->a1 [arrowhead=none]
  a->a1 [arrowhead=none]
  a1->b
  b->b1 [arrowhead=none]
  b1->c
  c->c1 [arrowhead=none]
  c1->d
  d->d1 [arrowhead=none]
  d1->e
  e->f; e->g
  f->g1 [arrowhead=none]; g->h1 [arrowhead=none]
  g1->h; h1->i 
  
  
{rank=same; b1->aa [minlen=2]}   
{rank=same; c1->bb [minlen=2]} 
{rank=same; d1->cc [minlen=2]}
{rank=same; g1->hh [minlen=2]}
{rank=same; h1->ii [minlen=2]}



#invisible constraints to force ordering
#ff->i [style=invis]
   }   
   
   [1]: n_fullFF
   [2]: n_methylQC
   [3]: n_methylFF
   [4]: n_withmaternaldata
   [5]: n_withchilddata
   [6]: n_Child
   [7]: n_Teen
   [8]: n_QCloss
   [9]: n_missingmaternalprenataldata
   [10]: n_missingchilddata
   [11]: n_ChildSmkexcluded
   [12]: n_TeenSmkexcluded
   [13]: n_ChildSmkexclusions
   [14]: n_TeenSmkexclusions
")

graph1

graph1 %>%
  export_svg() %>%
  read_xml() %>%
  write_xml(file.path(outdir, paste0("Figure", main_fig, ".svg")))

main_fig=main_fig+1            
```

# Supplemental Table 1 - Descriptive statistics of excluded vs included samples with methylation data available from the Fragile and Families and Child Wellbeing study

```{r}
myFF$MethylData2<-ifelse(myFF$MethID %in% modeldata$MethID, 'Final analysis group', 'Not in final analysis group')
set_label(myFF$MethylData2)<-'Exclusions based on missing data'
myFF=left_join(myFF, methyldata%>%mutate(idnum=as.character(idnum)))
myFF=myFF%>%mutate(across(c(PostnatalMaternalSmokingAny, SmkAtVisitPastmonth), ~ifelse(.x=='Missing', 
                                                       NA, .x)))%>%
  mutate(ancestry=ifelse(ancestry=='Missing PC data', NA, ancestry))%>%copy_labels(myFF)

exclusion_tbl=myFF %>% mutate(childteen=case_when(childteen=='C'~'Age 9', childteen=='T'~'Age 15'))%>%
  dplyr::select(smkPreg_binary, m1g2_YesNoPreg, m1g3_YesNoPreg, cm1inpov, PostnatalMaternalSmokingAny,
         SmkAtVisitPastmonth, ancestry, cm1bsex, childteen, globalmethylation, pediatric, grim,
         Epithelial.cells_saliva, Leukocytes_saliva, SSnewbornCT_center, SSolder_center,
         cg05575921, MethylData2)%>%
  tbl_summary(., by=MethylData2, statistic=list(all_continuous()~"{mean} ({sd})"))%>%
  add_n()%>%add_p(test=list(all_continuous()~"t.test"))

exclusion_tbl
```



# Supplemental Figure 2 - Correlation of methylation summary measures

```{r, fig.width=12, fig.height=12}
library(Hmisc)
library(corrplot)
res2<-modeldata%>%dplyr::select(any_of(all_methyl))%>%cor()
#labels
colnames(res2)=all_methyl_labels
rownames(res2)=all_methyl_labels
corrplot.mixed(res2, upper="ellipse", lower='number', order='hclust', tl.pos ="lt", tl.cex=0.8)

pdf(file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
corrplot.mixed(res2, upper="ellipse", lower='number', order='hclust', tl.pos ="lt", tl.cex=0.8)
dev.off()
sup_fig=sup_fig+1
    
```


# Supplemental Figure 3 - Correlation between visits among DNA methylation summary measures 

```{r}
age_long_methyl=modeldata%>%
  dplyr::select(any_of(c('idnum', 'childteen', 'ancestry', 'smkPreg_binary', all_methyl)))%>%
  pivot_longer(cols=any_of(all_methyl),
               names_to='Methylation summary measure type',
               values_to='Methylation summary measure')%>%
  mutate(Methyl_factor=factor(`Methylation summary measure type`, 
                              levels=all_methyl_levels))

age_wide_methyl=age_long_methyl%>%
  pivot_wider(names_from=childteen, 
              values_from=`Methylation summary measure`)

visit_cor=age_wide_methyl%>%
  ggplot(aes(x=`Age 9`, y=`Age 15`))+geom_point()+stat_cor()+geom_smooth(method=lm)+theme_classic()+
  scale_x_continuous('Age 9 score')+scale_y_continuous('Age 15 score')+
  facet_wrap(~Methyl_factor, scales='free', labeller = labeller(Methyl_factor=all_methyl_labels))

visit_cor

ggexport(visit_cor, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1
```

# Supplemental Figure XX: Distributions of methylation summary measures at age 9 and age 15

```{r, fig.height=11}
dist_methyl=age_long_methyl%>%
  ggplot(aes(x=childteen, y=`Methylation summary measure`, color=childteen))+
  geom_violin()+geom_boxplot(width=0.2)+stat_compare_means(label="..p.signif..", label.y.npc=0.87)+
  scale_color_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  scale_x_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  geom_point(alpha=0)+geom_line(alpha=0, aes(group=idnum))+geom_smooth(aes(group=1), method='lm', color='black')+
  facet_wrap(~Methyl_factor, scales='free', labeller = labeller(Methyl_factor=all_methyl_labels))+
  theme(legend.position = 'bottom')

dist_methyl

ggexport(dist_methyl, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1

```


# Figure 2: Polymethylation scores are correlated and stable between age 9 and age 15 in a diverse sample of `r n_total` US children 

```{r change in scores time, fig.width=12}

#add line for mean ? 
outcome_labels=c('Global methylation', 'Pediatric clock', 'AHHR: cg05575921', 'Polymethylation score \n (newborn, mean centered)', 'Polymethylation score  \n (older children, mean centered')
names(outcome_labels)=c('globalmethylation', 'pediatric', 'cg05575921', 'SSnewbornCT_center','SSolder_center')

fig2=age_long_methyl%>%
    filter(Methyl_factor%in% main_methyl)%>%
    ggplot(aes(x=childteen, y=`Methylation summary measure`, color=smkPreg_binary))+
    geom_violin()+geom_boxplot(width=0.2, position = position_dodge(0.9))+
  stat_compare_means(label='..p.signif..', label.y.npc = 0.88)+
    facet_wrap(~Methyl_factor, scales='free',labeller = labeller(Methyl_factor=all_methyl_labels))+
    scale_x_discrete('Visit', labels=c('Age 9', 'Age 15'))+
    scale_color_grey('', labels=c('Not prenatal maternal smoke exposed', 'Prenatal smoke exposed'))+
    theme(legend.position = 'bottom')


fig2

ggexport(fig2, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')))
main_fig=main_fig+1

```


# Table 1: Bivariate associations between prenatal maternal smoking, DNA methylation summary measures and important covariates among a diverse sample of 

```{r bivariate table}
smoking_tbl=modeldata %>% 
  dplyr::select(smkPreg_binary, m1g2_YesNoPreg, m1g3_YesNoPreg, cm1inpov, PostnatalMaternalSmokingAny,
         SmkAtVisitPastmonth, ancestry, cm1bsex, globalmethylation, pediatric, Epithelial.cells_saliva,
         Leukocytes_saliva, SSnewbornCT_center, SSolder_center, cg05575921, childteen)%>%
  tbl_strata(strata=childteen, .tbl_fun=~.x %>%
               tbl_summary(by=smkPreg_binary, statistic=
                             list(all_continuous()~"{mean} ({sd})"))%>%
               add_n()%>%add_p(test=list(all_continuous()~"t.test")))

smoking_tbl=smoking_tbl%>%as_gt()%>%
  tab_row_group(label='Maternal', rows = 1:11)%>%
  tab_row_group(label='Child', rows=12:25)

smoking_tbl
```


# Figure 3 - Polymethylation scores are associated with prenatal maternal smoke exposure consistently across age and ancestry group in a sample of `r n_total` US children

```{r, fig.width=12}
my_outcomes=c('Global methylation', 'Pediatric clock', 'AHHR: cg05575921', 'Sustained smoking polymethylation score (newborns, cell-type controlled)')

outcome_labels=c('Global methylation', 'Pediatric clock', 'AHHR: cg05575921',  'Polymethylation score \n (newborn, mean centered)')

all_models=rbind(local_models, global_models%>%mutate(ancestry='Unstratified'))

data_text=all_models%>%
  mutate(n=map_dbl(data, nrow), n=paste0('n=', n), x=-0.25)%>%
  unnest(tidied)%>%
  filter(outcome=='Sustained smoking polymethylation score (newborns, cell-type controlled)' & term=='smkPreg_binaryYes'& predictors=='secondhand smoke exposure model')%>%mutate(outcome_f=outcome)
  
names(outcome_labels)=my_outcomes

forest_plot=all_models%>%
  mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
  unnest(tidied)%>%
  filter(term=='smkPreg_binaryYes' & predictors=='secondhand smoke exposure model')%>%
  filter(outcome%in% my_outcomes)%>%
    mutate(outcome_f=factor(outcome, levels=rev(my_outcomes)))%>%
  ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, rev(levels(age))), y=ancestry))+
  geom_point(position = position_dodge(width = 1))+
  geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
  facet_grid(.~outcome_f, scales='free_x', labeller = labeller(outcome_f=outcome_labels))+
  geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
  geom_text(data=data_text, aes(x=x,label=n), position=position_dodge(width = 1), hjust=0.05)+
  theme_classic()+
  theme(text = element_text(size=16))+theme(legend.position = 'bottom')+
  scale_x_continuous('Coefficient for prenatal maternal smoking + 95% CI')+
  scale_y_discrete('', limits=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'), labels=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'))+scale_color_manual('', values=c('black', 'darkgrey'))

forest_plot

ggexport(forest_plot, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=15)
main_fig=main_fig+1
```

# Internal Table and Figure: Question: Remove postnatal maternal smoking variable from models and/or use base model as main paper figures? 

Somewhat concerned about extrapolated inference: very few exposed children (post-ancestry stratification) who were not also exposed to postnatal smoking. 

```{r}
with(modeldata, table(ancestry, PostnatalMaternalSmokingAny, smkPreg_binary))%>%kbl()

forest_plot_compareModels=all_models%>%
    mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
    unnest(tidied)%>%
    filter(term=='smkPreg_binaryYes')%>%
    filter(outcome%in% my_outcomes)%>%
    mutate(outcome_f=factor(outcome, levels=rev(my_outcomes)))%>%
    ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, rev(levels(age))), y=ancestry))+
    geom_point(position = position_dodge(width = 1))+
    geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
    facet_grid(predictors~outcome_f, scales='free_x', labeller = labeller(outcome_f=outcome_labels))+
    geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
    theme_classic()+
    theme(text = element_text(size=16))+theme(legend.position = 'bottom')+
    scale_x_continuous('Coefficient for prenatal maternal smoking + 95% CI')+
    scale_y_discrete('', limits=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'), labels=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'))+scale_color_manual('', values=c('black', 'darkgrey'))

forest_plot_compareModels

ggexport(forest_plot_compareModels, filename=file.path(outdir, paste0('InternalFigure', '1', '.pdf')), width=15, height=15)
```


# Supplemental Table XX: Associations between prenatal maternal smoking and DNA methylation summary measures in multivariable models among `r n_total` US children by age 

```{r}
#my_main_outcomes=c('AHHR: cg05575921', 'Sustained smoking polymethylation score (newborns, cell-type controlled)', 'Sustained smoking polymethylation score (older children)', 'Pediatric clock', 'Global methylation')

test=rbind(global_models, longitudinal_global_models %>% mutate(childteen='Longitudinal'))%>%
  filter(predictors=='secondhand smoke exposure model')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(childteen=factor(case_when(childteen=='Age 9'~'Age 9 models', 
                             childteen=='Age 15'~'Age 15 models', 
                             childteen=='Longitudinal'~'Longitudinal models'),levels=c('Age 9 models', 'Age 15 models', 'Longitudinal models')), 
         tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$childteen)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_age=tbl_stack(table_list, group_header=names(table_list))

regression_table_age

```

# Supplemental Table XX: Associations between prenatal maternal smoking and DNA methylation summary measures in multivariable models among `r n_total` US children by ancestry 


```{r}
test=rbind(longitudinal_global_models %>% mutate(ancestry='Unstratified'),
           longitudinal_local_models%>%dplyr::select(-data))%>%
  filter(predictors=='secondhand smoke exposure model')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$ancestry)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_ancestry=tbl_stack(table_list, group_header=names(table_list))


regression_table_ancestry

```





# Supplemental Figure XX: Surrogate variable associations with known covariates

```{r, fig.width=12, fig.height=8}
#bring the heatmap function
covar=c('Sample_Plate', 'Slide', 'Array', 'cm1bsex', 'smkPreg_binary', 'childteen', 'ancestry')

bring.the.heatmap.sv <- function(pd,covar,folder){
  
  nsv <- length(grep('sv',names(pd)))
  
  #generate correlation matrix between SVs and variables
  correlations <- matrix(0,nrow=length(covar),ncol=nsv)
  colnames(correlations) <- paste('sv',1:nsv,sep='')
  rownames(correlations) <- covar
  
  #take svs
  svs <- pd[,colnames(correlations)]
  
  #correlations between svs and variables:
  # cor test if continuous
  # anova is multi category
  # t-test if two category
  for(col in covar){
    var <- pd[,col]
    #check to see if it is numeric and isn't really a two category thing
    if(class(var)=='numeric' & dim(table(var))>2){
      correlations[col,] <- apply(svs,2,FUN = function(x,y) cor.test(x,y)$p.value,y=var)
    }else{
      if(length(levels(factor(var)))>2){
        correlations[col,] <- apply(svs,2,FUN = function(x,y) anova(lm(x~y))$'Pr(>F)'[1],y=var)
      }else{
        correlations[col,] <- apply(svs,2,FUN = function(x,y) t.test(x~y)$p.value,y=var)
      }
    }
  }
  color.gradient <- colorRampPalette(c('firebrick','ivory3','royalblue'),bias=3)
  correlations=correlations%>%as.data.frame()%>%tibble::rownames_to_column('var1')%>%
    pivot_longer(-var1, names_to='var2')%>%
    mutate(var2=factor(var2, levels=unique(mixedsort(var2))),
           var1=factor(var1, levels=pd[, covar]%>%colnames(), labels=get_label(pd[, covar])))
  correlations%>%
    ggplot(aes(x=var1, y=var2, fill=value))+geom_tile()+
    scale_fill_gradientn('P value', colours = color.gradient(20), values = seq(0, 1, 0.5))+
    scale_x_discrete('', labels=str_wrap(get_label(pd[, covar]), 10))+
    scale_y_discrete('')
}


bring.the.heatmap.sv(modeldata, covar=c('smkPreg_binary', 'Sample_Plate', 'Slide', 'Array', 'ChildAgeComposite', 'ancestry', 'cm1bsex', 'Leukocytes_saliva', 'Epithelial.cells_saliva', 'cm1inpov', 'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'PostnatalMaternalSmokingAny', 'SmkAtVisitPastmonth'))

pdf(file=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=11, height=13)
bring.the.heatmap.sv(modeldata, covar=c('smkPreg_binary', 'Sample_Plate', 'Slide', 'Array', 'ChildAgeComposite', 'ancestry', 'cm1bsex', 'Leukocytes_saliva', 'Epithelial.cells_saliva', 'cm1inpov', 'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'PostnatalMaternalSmokingAny', 'SmkAtVisitPastmonth'))
dev.off()
sup_fig=sup_fig+1
```


# Supplemental Table XX: Surrogate variable models 

```{r}
test=rbind(global_models, longitudinal_global_models %>% mutate(childteen='Longitudinal'))%>%
  filter(predictors=='surrogate variable models')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(childteen=factor(case_when(childteen=='Age 9'~'Age 9 models', 
                             childteen=='Age 15'~'Age 15 models', 
                             childteen=='Longitudinal'~'Longitudinal models'),levels=c('Age 9 models', 'Age 15 models', 'Longitudinal models')), 
         tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$childteen)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_age_svs=tbl_stack(table_list, group_header=names(table_list))


test=rbind(longitudinal_global_models %>% mutate(ancestry='Unstratified'),
           longitudinal_local_models%>%dplyr::select(-data))%>%
  filter(predictors=='surrogate variable models')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$ancestry)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_ancestry_svs=tbl_stack(table_list, group_header=names(table_list))

regression_table_ancestry_svs
```


# Figure 4 - Polymethylation scores more accurately delineate prenatal smoke exposure status than non-targeted DNA methylation summary measures and *a priori* CpGs 

```{r, fig.width=10}

age_labels=c('Age 9', 'Age 15'); names(age_labels)=c('Age 9', 'Age 15')
methyl_labels_roc=paste0('Base model+', all_methyl_labels)
names(methyl_labels_roc)=paste0('Base model+', names(all_methyl_labels))

fig4a= global_roc%>%
    dplyr::select(predictors, covariates, methylation, childteen, test, modeltype)%>%
    unnest(test) %>% 
    filter(str_detect(methylation, 'SSnewbornCT|cg0557|global') | modeltype=='Base model')%>%
    ggplot(aes(x=spec, y=sens, color=modeltype))+geom_line()+
    facet_wrap(childteen~., nrow=1, labeller=labeller(childteen=age_labels))+
    theme_classic()+
    scale_color_manual('', values=c('red', 'darkred', 'darkblue', 'darkgreen', 'blue', 'green'), 
                       function(x) str_wrap(x, width = 30))+
    theme(legend.position = 'right')+guides(color=guide_legend(nrow=4, byrow=TRUE))+
    geom_abline(intercept=1, slope=-1, color='black')+
    xlab('Specificity')+ylab('Sensitivity')+
    theme(legend.position='bottom')+ coord_fixed()

fig4b=global_roc%>%
  dplyr::select(covariates, methylation, childteen, test)%>% 
  filter(str_detect(methylation, 'SSnewbornCT') & covariates!='')%>%unnest(test)%>%
  rbind(long_global_roc%>% 
          mutate(childteen='Age 9 & Age 15', methylation=gsub('_Age_9.*', '', methylation))%>%
          dplyr::select(covariates, methylation, childteen, test)%>% 
          filter(str_detect(methylation, 'SSnewbornCT') 
                 & covariates!='')%>%
          unnest(test))%>%
  ggplot(aes(x=spec, y=sens, color=childteen, linetype=childteen))+geom_line(size=2)+
  facet_wrap(~methylation, labeller = labeller(methylation=c('SSnewbornCT_center'='Polymethylation score, \n newborn mean centered')))+
  theme_classic()+theme(legend.position = 'right')+guides(color=guide_legend(nrow=4,byrow=TRUE))+
  geom_abline(intercept=1, slope=-1, color='black')+
  scale_color_grey('', labels=c('Age 9 only', 'Age 9 & 15', 'Age 15 only'))+
  scale_linetype('', labels=c('Age 9 only', 'Age 9 & 15', 'Age 15 only'))+
  xlab('Specificity')+ylab('Sensitivity')+
  theme(legend.position='bottom')+ coord_fixed()

auc_fig=cowplot::plot_grid(fig4a, fig4b, labels=c('A', 'B'), nrow=1, rel_widths=c(0.65, 0.35), align='h', axis='t')

auc_fig

ggexport(auc_fig, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=9)
main_fig=main_fig+1
```

# Supplemental Table with AUC and p-value comparison to sustained smoking, newborn cell type controls polymethylation score

```{r}
roc_table=global_roc %>% dplyr::select(childteen, methylation, modeltype, roc, auc)
roc_base=roc_table %>% filter(modeltype=='Base model+No methylation')%>%mutate(roc_basemodel=roc)%>%dplyr::select(childteen, roc_basemodel)
roc_ss=roc_table %>% filter(modeltype=='Sustained smoking w/ cell type correction (newborn)')%>%mutate(roc_pms=roc)%>%dplyr::select(childteen, roc_pms)

roc_table1=roc_table%>%filter(!str_detect(modeltype, 'Base model'))%>%left_join(roc_ss)%>%mutate(roctest=map2(roc, roc_pms, ~roc.test(.x, .y)))%>%mutate(auc=round(auc, 2))

roc_table1%>%mutate(`pval comparison to sustained smoking score (newborns)`=map_dbl(roctest, ~round(.x$p.value, 4)))%>%dplyr::select(childteen, modeltype, auc, `pval comparison to sustained smoking score (newborns)`)%>%kbl(col.names=c('Age', 'Model', 'AUC', 'P value (vs polymethylation score, newborns, cell type controlled)'))
```

# Supplemental Table with AUC and p-value comparison to base model with no methylation variables (i.e., using only child sex, income to poverty ratio, saliva/epithelial cell proportions and sample plate)

```{r}

roc_table2=roc_table%>%filter(str_detect(modeltype, 'Base model'))%>%left_join(roc_base)%>%mutate(roctest=map2(roc, roc_basemodel, ~roc.test(.x, .y)))%>%mutate(auc=round(auc, 2))

roc_table2%>%mutate(`pval comparison to base model`=map_dbl(roctest, ~round(.x$p.value, 4)))%>%dplyr::select(childteen, modeltype, auc, `pval comparison to base model`)%>%kbl(col.names=c('Age', 'Model', 'AUC', 'P value (vs base w/ no methyl)'))


```

