---
title: "Fragile Families - Epigenome wide association study"
author: "Freida Blostein"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    code_folding: show
    highlight: tango
    number_sections: yes
    theme: sandstone
    toc: yes
    toc_depth: 2
    toc_float: yes
---

```{r setup, include=FALSE, eval=F}
if (!requireNamespace("BiocManager")) install.packages("Bioconductor")

BiocManager::install(c(
  "bumphunter", # Differentially methylated region analysis
  "minfi", # Many functions for methylation analysis
  "sva", # Batch effect correction
  "limma", # Single-site association analysis
  "IlluminaHumanMethylation450kmanifest", # Manifest file
  "IlluminaHumanMethylation450kanno.ilmn12.hg19", # Annotation file
  "GO.dv", # Gene ontology queries
  # "missMethyl",
  "GEOquery"
))

install.packages(c(
  "RColorBrewer", # Color palettes for data visualization,
  "ggplot2", # Useful and aesthetic plotting functions
  "MatrixEQTL", # Fast eQTL and meQTL analyses
  "qqman", # QQ and manhattan plots
  "nlme", # Non-linear mixed effects models
  "BiasedUrn" # Hypergeometric distributions
))
```

```{r load libraries, include=F}
knitr::opts_chunk$set(echo = TRUE)
library(minfi)
library(limma)
library(matrixStats)
library(MASS)
#library(sva) --> current problem with png package, lib.png.16 https://stackoverflow.com/questions/47424249/unable-to-install-the-package-png-on-r-possible-conflict-with-anaconda/50935430
#library(Hmisc)
#library( missMethyl )
library(ggplot2)
library(qqman)
library(tidyr)
library(dplyr)
library(ggupset)
library(purrr)
library(broom.mixed)
library(future)
library(furrr)
library(tibble)
library(kableExtra)
library(cowplot)
library(matrixStats)
library(lme4)
library(lmerTest)
library(ggpubr)
data.dir="/scratch/bfoxman_root/bfoxman/blostein/FragileFamilies/Files/"
limma.out.dir<-file.path(datadir, 'CreatedData', 'limmaModels')
outputdir<-"/scratch/bfoxman_root/bfoxman/blostein/FragileFamilies/Output"
```

```{r read in data}
#load in betaqc matrix
#betaqc<-readRDS(file=paste0(data.dir, 'OGData/', "noob_filtered.rds"))
#transform to m values in case want to use
#m_val<-function(x){log2(x/(1-x))}
#mvaluesqc<-apply(betaqc, 2, m_val)

#load in metadata - child no smoke and teen no smoke
load(file=paste0(data.dir, 'CreatedData/', 'ChildNoSmkModelData.Rdata'))
load(file=paste0(data.dir, 'CreatedData/', 'TeenNoSmkModelData.Rdata'))
#drop missing factor levels for both data sets
child_nosmoke<-child_nosmoke%>%droplevels()
teen_nosmoke<-teen_nosmoke%>%droplevels()
#set factor levels
child_nosmoke$PostnatalMaternalSmokingAny<-factor(child_nosmoke$PostnatalMaternalSmokingAny, levels=c("No maternal smoking at age 1 and age 5", "Maternal smoking at age 1 or age 5"))
child_nosmoke$SmkAtVisitPastmonth<-factor(child_nosmoke$SmkAtVisitPastmonth, levels=c("No smoking", "Less than pack a day", "Pack or more a day"))
teen_nosmoke$PostnatalMaternalSmokingAny<-factor(teen_nosmoke$PostnatalMaternalSmokingAny, levels=c("No maternal smoking at age 1 and age 5", "Maternal smoking at age 1 or age 5"))
teen_nosmoke$SmkAtVisitPastmonth<-factor(teen_nosmoke$SmkAtVisitPastmonth, levels=c("No smoking", "Less than pack a day", "Pack or more a day"))
modeldata=rbind(child_nosmoke, teen_nosmoke)
```

# Preprocessing

```{r filter invariant probes}
#probeVars<-rowVars(betaqc)
#saveRDS(probeVars, file=paste0(data.dir, 'CreatedData/', "probevariance.rds"))
probeVars<-readRDS(file=paste0(data.dir, 'CreatedData/', "probevariance.rds"))
varCut<-0.0002
a<-ggplot()+aes(probeVars)+geom_histogram(boundary=0)+theme_classic2()+ggtitle('Histogram of probe variances')+scale_x_continuous('Probe variances', breaks=seq(0, 0.15, 0.025))+annotate('rect', xmin=0, xmax=0.01, ymin=0, ymax=4e5, fill=alpha('grey', 0), color='orange')+annotate('segment', x=0.01, xend=0.04, y=1e5, yend=2e5, arrow=arrow(), color='orange')

b<-ggplot()+aes(probeVars[probeVars<0.01])+geom_histogram()+theme_classic2()+ggtitle('Histogram of probe variances<0.01')+scale_x_continuous('Probe variances', breaks=seq(0, 0.01, 0.005))+geom_vline(xintercept = varCut, color='red', linetype='dashed')+annotate('text', x=0.003, y=1e05, label=paste0('Cut ', length(probeVars[probeVars<varCut]), ' invariant probes \n', 'using variance cut of >', varCut), color='red')

#betaqc<-betaqc[rowVars(betaqc)>varCut,]
#saveRDS(betaqc, file=paste0(data.dir, 'CreatedData/', 'betaTrim.Rds'))
betaqc<-readRDS(file=paste0(data.dir, 'CreatedData/', 'betaTrim.Rds'))
```

```{r draw probe variance histogram, fig.width=8}
ggdraw(a)+draw_plot(b, 0.35, 0.35, 0.55, 0.55)
```

Based on the above histogram, I select `r varCut` as the lower limit for probe variance; all probes with variances<`r varCut` were removed from the set ($m_{CpG}$=`r length(probeVars[probeVars<varCut])` removed), leaving a total of $m_{CpG}$=`r nrow(betaqc)` probes for the EWAS. 

```{r check variance filter dimensions}
if(!(nrow(betaqc)+length(probeVars[probeVars<varCut])==length(probeVars))){print('STOP: invariant probe filter cut before vs after probe count incorrect')}
```

```{r annotation}
### append annotation information
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data(Locations)
data(Islands.UCSC)
data(Other)
pos<-Locations[rownames(betaqc),'pos']
chrnames <- as.character(Locations[rownames(betaqc),'chr'])

add_annotations<-function(data){
  #add annotations
Locations <- Locations[rownames(data), ]
Islands.UCSC <- Islands.UCSC[rownames(data),]
Other <- Other[rownames(data),]
# add columns of interest from annotation to annotated version
data.annotated <- data
data.annotated$chr <- Locations$chr
data.annotated$pos <- Locations$pos
data.annotated$island <- Islands.UCSC$Relation_to_Island
data.annotated$gene <- Other$UCSC_RefGene_Name
data.annotated$regulatory_feature <- Other$Regulatory_Feature_Group
return(data.annotated)
}

```

# Cross sectional epigenome wide association study: Children Age 9

```{r beta filtering and checks}
#filter betaqc to child samples only 
beta_c <-betaqc[, colnames(betaqc) %in% child_nosmoke$MethID]#; rm(betaqc)
#make sure that the order of the samples matches in the metadata matrix and the beta matrix are ordered the same
child_nosmoke<-child_nosmoke[match(colnames(beta_c), child_nosmoke$MethID), ]
#checks
if(dim(beta_c)[2]!=dim(child_nosmoke)[1]){print('STOP: Sample #s do not match')}
if(identical(colnames(beta_c), child_nosmoke$MethID)==FALSE){print('STOP: Sample names do not match')}
```

```{r child cross sectional ewas global model1, eval=F}
#model 1: make model matrix
mod1=model.matrix(~factor(child_nosmoke$smkPreg_binary)+child_nosmoke$global_PC1+child_nosmoke$global_PC2+factor(child_nosmoke$cm1bsex)+child_nosmoke$cm1inpov+child_nosmoke$ChildAgeComposite+child_nosmoke$Epi+child_nosmoke$IC)

#run EWAS with empirical st bayes approach to reduce standard error
out<-lmFit(beta_c, mod1)
out<-eBayes(out)
saveRDS(out, file=file.path(limma.out.dir, 'limmaModel1.Rda'))

#pull out top hits
child.ss.hits<-topTable(out, coef=2, number=nrow(beta_c), confint = T)
rm(mod1, out)

# Annotate top hits
child.m1.ss.hits.annotated <- add_annotations(child.ss.hits)
saveRDS(child.m1.ss.hits.annotated, file=file.path(limma.out.dir, 'Model1TopHits.Rda'))
```

```{r load model1}
child.m1.ss.hits.annotated <-readRDS(file=file.path(limma.out.dir, 'Model1TopHits.Rda'))
```

```{r child cross sectional ewas global model2, eval=F}
#model 2: make model matrix
mod2=model.matrix(~factor(child_nosmoke$smkPreg_binary)+child_nosmoke$global_PC1+child_nosmoke$global_PC2+factor(child_nosmoke$cm1bsex)+child_nosmoke$cm1inpov+child_nosmoke$ChildAgeComposite+child_nosmoke$Epi+child_nosmoke$IC+factor(child_nosmoke$m1g2_YesNoPreg)+factor(child_nosmoke$m1g3_YesNoPreg))

#run EWAS with empirical st bayes approach to reduce standard error
out<-lmFit(beta_c, mod2)
out<-eBayes(out)
saveRDS(out, file=file.path(limma.out.dir, 'limmaModel2.Rda'))

#pull out top hits
child.ss.hits<-topTable(out, coef=2, number=nrow(beta_c))
rm(mod2, out)

#annotate
child.m2.ss.hits.annotated <- add_annotations(child.ss.hits)
saveRDS(child.m2.ss.hits.annotated, file=file.path(limma.out.dir, 'Model2TopHits.Rda'))
```

```{r load model2}
child.m2.ss.hits.annotated <-readRDS(file=file.path(limma.out.dir, 'Model2TopHits.Rda'))
```

```{r child cross sectional ewas global model3, eval=F}
#model 3: make model matrix
mod3=model.matrix(~factor(child_nosmoke$smkPreg_binary)+child_nosmoke$global_PC1+child_nosmoke$global_PC2+factor(child_nosmoke$cm1bsex)+child_nosmoke$cm1inpov+child_nosmoke$ChildAgeComposite+child_nosmoke$Epi+child_nosmoke$IC+factor(child_nosmoke$m1g2_YesNoPreg)+factor(child_nosmoke$m1g3_YesNoPreg)+factor(child_nosmoke$PostnatalMaternalSmokingAny)+factor(child_nosmoke$SmkAtVisitPastmonth))

#run EWAS with empirical st bayes approach to reduce standard error
out<-lmFit(beta_c, mod3)
out<-eBayes(out)
saveRDS(out, file=file.path(limma.out.dir, 'limmaModel3.Rda'))
#pull out top hits
child.ss.hits<-topTable(out, coef=2, number=nrow(beta_c))
rm(mod3, out)


#annotate
child.m3.ss.hits.annotated <- add_annotations(child.ss.hits)
saveRDS(child.m3.ss.hits.annotated, file=file.path(limma.out.dir, 'Model3TopHits.Rda'))
```

```{r load model3}
child.m3.ss.hits.annotated <-readRDS(file=file.path(limma.out.dir, 'Model3TopHits.Rda'))
```

### Future analysis: compare top CpGs when run in ancestry specific models

## QQ plots

```{r calculate lambdas}
calcLambda<-function(df){
  o<- -log10(sort(df$P.Value, decreasing = F))
  e<- -log10(ppoints(length(df$P.Value)))
  lambda<-median(o)/median(e)
  return(lambda)
}

l_c1<-calcLambda(child.m1.ss.hits.annotated)
l_c2<-calcLambda(child.m2.ss.hits.annotated)
l_c3<-calcLambda(child.m3.ss.hits.annotated)

```

```{r QQPlots, fig.height=16}
par(mfrow=c(3, 1), mar=c(5.1, 4.1, 5.1, 2.1))
qq(child.m1.ss.hits.annotated$P.Value, main='Base model: prenatal smoking, genetic PCs, sex \n age, income to poverty ratio, Epi, Fib')
mtext(paste0(expression(lambda), '=', round(l_c1, 3)), side=3)
qq(child.m2.ss.hits.annotated$P.Value, main='Prenatal exposures model: \n Base model vars+ prenatal alcohol+prenatal other drug')
mtext(paste0(expression(lambda), '=', round(l_c2, 3)), side=3)
qq(child.m3.ss.hits.annotated$P.Value, main='Secondhand smoking model:  Prenatal exposure model + postnatal maternal smoking any at 1 or 5 + \n postnatal maternal smoking pks/day in month prior to visit')
mtext(paste0(expression(lambda), '=', round(l_c3, 3)), side=3)
```


## Manhattan plot

```{r Manhattan plots}
makeManhattan<-function(ss.hits, beta){
  pos.ss <- pos[match(rownames(ss.hits), rownames(beta))]
  chr.ss <- chrnames[match(rownames(ss.hits), rownames(beta))]
  chr.ss.num <- as.numeric(unlist(lapply(chr.ss, function(x) {
    strsplit(x, "chr", fixed = TRUE)[[1]][2]
  })))
  
  # Construct data frame for function
  formanhattan <- data.frame(pos.ss, chr.ss.num, ss.hits$P.Value)
  colnames(formanhattan) <- c("BP", "CHR", "P")
  
  formanhattan <- formanhattan[-which(is.na(formanhattan$CHR)), ]
  return(formanhattan)
}

m1.forman<-makeManhattan(child.m1.ss.hits.annotated, beta_c)
m2.forman<-makeManhattan(child.m2.ss.hits.annotated, beta_c)
m3.forman<-makeManhattan(child.m3.ss.hits.annotated, beta_c)

```

### Manhattan in children (Age 9) model 1

```{r manhattan model1}
manhattan(m1.forman)
```

## Beta vs Beta plots within Age 9 models

```{r beta vs beta within child}
child_models<-list('child_model1'=child.m1.ss.hits.annotated, 'child_model2'=child.m2.ss.hits.annotated, 'child_model3'=child.m3.ss.hits.annotated)%>%lapply(., function(x) rownames_to_column(x, var='CpG'))%>%Map(cbind, ., ModelType=names(.))

beta_child<-lapply(child_models, function(x) x %>%dplyr::select(CpG, logFC, ModelType)%>%pivot_wider(id_cols = CpG, values_from=logFC, names_from=ModelType))%>%purrr::reduce(left_join, by='CpG')
  
#bind_rows(.id='ModelType')%>%mutate('Model'=gsub('.*_', '', ModelType), 'Age'=gsub('_.*', '', ModelType))

sig_beta_m1<-child_models[[1]]%>%filter(P.Value<1e-3)%>%pull(CpG)#change this to 1

smooth_betas<-function(df, x, y){
  x<-enquo(x); y<-enquo(y)
  ggplot(df, aes(x=!!x, y=!!y))+
    geom_point()+geom_smooth()+
    stat_cor()+theme_classic()+
    scale_y_continuous(name=paste0(quo_name(y)))+
    scale_x_continuous(name=paste0(quo_name(x)))
}

a<-smooth_betas(beta_child, child_model1, child_model2)
b<-smooth_betas(beta_child, child_model1, child_model3)
c<-smooth_betas(beta_child, child_model2, child_model3)

d<-smooth_betas(beta_child%>%filter(CpG %in% sig_beta_m1), child_model1, child_model2)
e<-smooth_betas(beta_child%>%filter(CpG %in% sig_beta_m1), child_model1, child_model3)
f<-smooth_betas(beta_child%>%filter(CpG %in% sig_beta_m1), child_model2, child_model3)

all_beta<-plot_grid(a, b, c, ncol=1)
sig_beta<-plot_grid(d, e, f, ncol=1)

```

```{r child beta vs beta draw graph, figure.height=16, fig.cap='Correlation between effect estimates from the three different epigenome wide linear models in children for A) all CpGs and B) significant CpGs only'}
plot_grid(all_beta, sig_beta, ncol=2, labels=c('A', 'B'), hjust = -2)

```


## Table of top hits in children models

### Model 1

```{r top hits ewas table child1}
child.m1.ss.hits.annotated%>% dplyr::rename(Beta.Coefficient=logFC, Average.Methylation=AveExpr)%>% head(n=10)%>%kable()%>%kable_classic()%>%scroll_box(width='100%', height='400px')
```

### Model 2

```{r top hits ewas table child2}
child.m2.ss.hits.annotated%>% dplyr::rename(Beta.Coefficient=logFC, Average.Methylation=AveExpr)%>%head(n=10)%>%kable()%>%kable_classic()%>%scroll_box(width='100%', height='400px')
```

### Model 3

```{r top hits ewas table child 3}
child.m3.ss.hits.annotated%>% dplyr::rename(Beta.Coefficient=logFC, Average.Methylation=AveExpr)%>%head(n=10)%>%kable()%>%kable_classic()%>%scroll_box(width='100%', height='400px')
```

```{r remove beta_c}
rm(beta_c)
```

# Cross sectional epigenome wide association study: Teen Age 15

```{r beta filtering and checks teens, eval=F}
#load in betaqc matrix
#betaqc<-readRDS(file=paste0(data.dir, 'OGData/', "noob_filtered.rds"))
#filter betaqc to child samples only 
beta_t <-betaqc[, colnames(betaqc) %in% teen_nosmoke$MethID]#; rm(betaqc)
#make sure that the order of the samples matches in the metadata matrix and the beta matrix are ordered the same
teen_nosmoke<-teen_nosmoke[match(colnames(beta_t), teen_nosmoke$MethID), ]
#checks
if(dim(beta_t)[2]!=dim(teen_nosmoke)[1]){print('STOP: Sample #s do not match')}
if(identical(colnames(beta_t), teen_nosmoke$MethID)==FALSE){print('STOP: Sample names do not match')}
```

```{r set models2, eval=F}
mod1=model.matrix(~factor(teen_nosmoke$smkPreg_binary)+teen_nosmoke$global_PC1+teen_nosmoke$global_PC2+factor(teen_nosmoke$cm1bsex)+teen_nosmoke$cm1inpov+teen_nosmoke$ChildAgeComposite+teen_nosmoke$Epi+teen_nosmoke$IC)

mod2=model.matrix(~factor(teen_nosmoke$smkPreg_binary)+teen_nosmoke$global_PC1+teen_nosmoke$global_PC2+factor(teen_nosmoke$cm1bsex)+teen_nosmoke$cm1inpov+teen_nosmoke$ChildAgeComposite+teen_nosmoke$Epi+teen_nosmoke$IC+factor(teen_nosmoke$m1g2_YesNoPreg)+factor(teen_nosmoke$m1g3_YesNoPreg))

mod3=model.matrix(~factor(teen_nosmoke$smkPreg_binary)+teen_nosmoke$global_PC1+teen_nosmoke$global_PC2+factor(teen_nosmoke$cm1bsex)+teen_nosmoke$cm1inpov+teen_nosmoke$ChildAgeComposite+teen_nosmoke$Epi+teen_nosmoke$IC+factor(teen_nosmoke$m1g2_YesNoPreg)+factor(teen_nosmoke$m1g3_YesNoPreg)+factor(teen_nosmoke$PostnatalMaternalSmokingAny)+factor(teen_nosmoke$SmkAtVisitPastmonth))

modellist=list(mod1, mod2, mod3)
modelnames=c('limmaModel1Teen.Rda', 'limmaModel2Teen.Rda', 'limmaModel3Teen.Rda')
tophitnames=c('TeenModel1TopHits.Rda', 'TeenModel2TopHits.Rda', 'TeenModel3TopHits.Rda')
```

```{r loop through models, eval=F}
for(i in 1:length(modellist)){
  print(paste('Running Model', i))
  out=lmFit(beta_t, modellist[[i]])
  out=eBayes(out)
  saveRDS(out, file=file.path(limma.out.dir, modelnames[i]))
  teen.ss.hits<-topTable(out, coef=2, number=nrow(beta_t), confint=T)
  rm(out)
  print(paste('Annotating top hits from model', i))
  teen.ss.annotated.hits<-add_annotations(teen.ss.hits)
  saveRDS(teen.ss.annotated.hits, file=file.path(limma.out.dir, tophitnames[i]))
  rm(teen.ss.hits, teen.ss.annotated.hits)
}
```

```{r load top hits in teen models}
teen.m1.ss.hits.annotated <-readRDS(file=file.path(limma.out.dir, 'TeenModel1TopHits.Rda'))
teen.m2.ss.hits.annotated <-readRDS(file=file.path(limma.out.dir, 'TeenModel2TopHits.Rda'))
teen.m3.ss.hits.annotated <-readRDS(file=file.path(limma.out.dir, 'TeenModel3TopHits.Rda'))
```

## QQ plots

```{r calc lambda teens}
l_t1<-calcLambda(teen.m1.ss.hits.annotated)
l_t2<-calcLambda(teen.m2.ss.hits.annotated)
l_t3<-calcLambda(teen.m3.ss.hits.annotated)

```

```{r QQPlots teens, fig.height=16}
par(mfrow=c(3, 1), mar=c(5.1, 4.1, 5.1, 2.1))
qq(teen.m1.ss.hits.annotated$P.Value, main='Base model: prenatal smoking, genetic PCs, sex \n age, income to poverty ratio, Epi, Fib')
mtext(paste0(expression(lambda), '=', round(l_t1, 3)), side=3)
qq(teen.m2.ss.hits.annotated$P.Value, main='Prenatal exposures model: \n Base model vars+ prenatal alcohol+prenatal other drug')
mtext(paste0(expression(lambda), '=', round(l_t2, 3)), side=3)
qq(teen.m3.ss.hits.annotated$P.Value, main='Secondhand smoking model: \n Prenatal exposure model + postnatal maternal smoking any at 1 or 5 + \n postnatal maternal smoking pks/day in month prior to visit')
mtext(paste0(expression(lambda), '=', round(l_t3, 3)), side=3)
```

## Beta vs Beta plots within Age 15 models

```{r beta vs beta within teen}
teen_models<-list('teen_model1'=teen.m1.ss.hits.annotated, 'teen_model2'=teen.m2.ss.hits.annotated, 'teen_model3'=teen.m3.ss.hits.annotated)%>%lapply(., function(x) rownames_to_column(x, var='CpG'))%>%Map(cbind, ., ModelType=names(.))

beta_teen<-lapply(teen_models, function(x) x %>%dplyr::select(CpG, logFC, ModelType)%>%pivot_wider(id_cols = CpG, values_from=logFC, names_from=ModelType))%>%purrr::reduce(left_join, by='CpG')

sig_beta_m1_t<-teen_models[[1]]%>%filter(P.Value<10e-4)%>%pull(CpG)

a<-smooth_betas(beta_teen, teen_model1, teen_model2)
b<-smooth_betas(beta_teen, teen_model1, teen_model3)
c<-smooth_betas(beta_teen, teen_model2, teen_model3)

d<-smooth_betas(beta_teen%>%filter(CpG %in% sig_beta_m1_t), teen_model1, teen_model2)
e<-smooth_betas(beta_teen%>%filter(CpG %in% sig_beta_m1_t), teen_model1, teen_model3)
f<-smooth_betas(beta_teen%>%filter(CpG %in% sig_beta_m1_t), teen_model2, teen_model3)

all_beta<-plot_grid(a, b, c, ncol=1)
sig_beta<-plot_grid(d, e, f, ncol=1)

```

```{r teen beta vs beta draw graph, figure.height=16, fig.cap='Correlation between effect estimates from the three different epigenome wide linear models in teens for A) all CpGs and B) significant CpGs only'}
plot_grid(all_beta, sig_beta, ncol=2, labels=c('A', 'B'), vjust = 0.8)
```

## Table of top hits in teen EWAS

### Model 1

```{r top hits ewas table teen1}
teen.m1.ss.hits.annotated%>% dplyr::rename(Beta.Coefficient=logFC, Average.Methylation=AveExpr)%>%head(n=10)%>%kable()%>%kable_classic()%>%scroll_box(width='100%', height='400px')
```

### Model 2

```{r top hits ewas table teen2}
teen.m2.ss.hits.annotated%>% dplyr::rename(Beta.Coefficient=logFC, Average.Methylation=AveExpr)%>%head(n=10)%>%kable()%>%kable_classic()%>%scroll_box(width='100%', height='400px')
```

### Model 3

```{r top hits ewas table teen3}
teen.m3.ss.hits.annotated%>% dplyr::rename(Beta.Coefficient=logFC, Average.Methylation=AveExpr)%>%head(n=10)%>%kable()%>%kable_classic()%>%scroll_box(width='100%', height='400px')
```

# Volcano plots for child and teen cross-sectional EWAS

```{r volcano}
volc_child<-ggplot(child.m1.ss.hits.annotated%>%mutate(adjPval=ifelse(adj.P.Val<0.05, '<0.05', ifelse(adj.P.Val<0.2, '<0.2', '>0.2'))), aes(x=logFC, y=-log10(P.Value), color=adjPval))+geom_point()+scale_x_continuous('Beta (regression coefficient)')#+geom_hline(yintercept=-log10(0.05/dim(child.m2.ss.hits.annotated)[1]))
volc_teen<-ggplot(teen.m1.ss.hits.annotated%>%mutate(adjPval=ifelse(adj.P.Val<0.05, '<0.05', ifelse(adj.P.Val<0.2, '<0.2', '>0.2'))), aes(x=logFC, y=-log10(P.Value), color=adjPval))+geom_point()+scale_x_continuous('Beta (regression coefficient)')#+geom_hline(yintercept=-log10(0.05/dim(child.m2.ss.hits.annotated)[1]))
```


```{r save volcano}
png(file.path(outputdir, 'Volcano.png'), width=1300, height=800)
plot_grid(volc_child+theme_classic()+theme(text=element_text(size=16)), volc_teen+theme_classic()+theme(text=element_text(size=16)), ncol=2, labels=c('A Child', 'B Teen'), vjust = 0.8, label_size = 18)
dev.off()
```


# Compare top hits in child and teen EWAS, and to other studies

```{r compare}
AllModels<-c(child_models, teen_models)



AllModelsdf=AllModelsdf%>%group_by(Model, Age, ModelType)%>%nest()%>%mutate(TopCpGs=map(data, function(x) x %>%filter(P.Value<1e-3)%>%pull(CpG)))%>%mutate(ModelType=case_when(ModelType=='child_model1'~'Age 9: Model 1', ModelType=='child_model2'~'Age 9: Model 2', ModelType=='child_model3'~'Age 9: Model 3', ModelType=='teen_model1'~'Age 15: Model 1', ModelType=='teen_model2'~'Age 15: Model 2', ModelType=='teen_model3'~'Age 15: Model 3'))%>%mutate(ModelType=factor(ModelType, levels=c('Age 9: Model 1', 'Age 9: Model 2', 'Age 9: Model 3', 'Age 15: Model 1', 'Age 15: Model 2', 'Age 15: Model 3')))
```

## Upset plot of top ten CpGs in each model:age 

```{r upset plot, fig.width=9}
a<-AllModelsdf %>% unnest(TopCpGs)%>%ungroup()%>%dplyr::select(ModelType, TopCpGs)%>%group_by(TopCpGs)%>%summarise(TypeList=list(ModelType))%>%ggplot(aes(x=TypeList))+geom_bar()+theme_classic2()+scale_x_upset(sets=c('Age 9: Model 1', 'Age 9: Model 2', 'Age 9: Model 3', 'Age 15: Model 1', 'Age 15: Model 2', 'Age 15: Model 3'), name='')+scale_y_continuous(name='')
b<-AllModelsdf %>%mutate(nCpGs=lengths(TopCpGs))%>%ggplot(aes(x=ModelType, y=-nCpGs))+geom_bar(position=position_dodge(), stat='identity')+coord_flip()+scale_y_continuous('',breaks=seq(0, -420, -50), labels=seq(0, 420, 50))+scale_x_discrete(limits=rev(levels(AllModelsdf$ModelType)))+theme_classic2()+theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y=element_blank(), axis.line.y = element_blank())
plot_grid(plot_grid(NULL, b + theme(plot.margin = unit(c(1, -5, -5, 1), "pt")), ncol = 1, rel_heights = c(2.15, 1)),a, nrow = 1, rel_widths = c(1, 3))
```

## Beta estimates comparison of top hits from child and teen models 

```{r betas over time}
beta_all<-lapply(AllModels, function(x) x %>%dplyr::select(CpG, logFC, ModelType)%>%pivot_wider(id_cols = CpG, values_from=logFC, names_from=ModelType))%>%purrr::reduce(left_join, by='CpG')

sig_beta_m1<-child_models[[1]]%>%filter(P.Value<10e-4)%>%pull(CpG)


a<-smooth_betas(beta_all, child_model1, teen_model1)
b<-smooth_betas(beta_all, child_model2, teen_model2)
c<-smooth_betas(beta_all, child_model3, teen_model3)

d<-smooth_betas(beta_all%>%filter(CpG %in% sig_beta_m1), child_model1, teen_model1)
e<-smooth_betas(beta_all%>%filter(CpG %in% sig_beta_m1), child_model2, teen_model2)
f<-smooth_betas(beta_all%>%filter(CpG %in% sig_beta_m1), child_model2, teen_model3)

all_beta<-plot_grid(a, b, c, ncol=1)
sig_beta<-plot_grid(d, e, f, ncol=1)
all_beta_pres<-plot_grid(a, b, c, ncol=3)
sig_beta_pres<-plot_grid(d, e, f, ncol=3)


```

```{r teen beta vs child beta draw graph, figure.height=16, , fig.cap='Correlation between effect estimates from the three different epigenome wide linear models in children vs in teens for A) all CpGs and B) significant CpGs only'}
plot_grid(all_beta, sig_beta, ncol=2, labels=c('A', 'B'), vjust = 0.8)
```

```{r save beta comparison}
png(file.path(outputdir, 'BetaChildvsTeen.png'), width=1300, height=800)
plot_grid(all_beta_pres+theme(text=element_text(size=16)), sig_beta_pres+theme(text=element_text(size=16)), ncol=1, labels=c('A', 'B'), vjust = 0.8)
dev.off()
```

## Similar top hits as Wicklund et al. 2019 and Joubert et al. 2016

```{r boxplots}
sigCpGs<-AllModelsdf%>%filter(Model=='model1')%>%mutate(data=map(data, function(x) x %>%filter(adj.P.Val<0.2)), SigCpG=map(data, function(x) x%>%pull(CpG)))%>%unnest(SigCpG)%>%pull(SigCpG) %>%unique()

betaqc<-readRDS(file=paste0(data.dir, 'CreatedData/', 'betaTrim.Rds'))
mybetaqc<-betaqc[sigCpGs%>%unique(), ]
#allgaps<-gaphunter(betaqc) #terminates r studio session (memory) 
mygaps<-gaphunter(mybetaqc)

myBeta<-as.data.frame(t(mybetaqc))
myBeta$MethID =row.names(myBeta)
modeldata2<-left_join(modeldata%>%dplyr::select(-contains('cg')), myBeta)

boxplotcpg<-modeldata2 %>% pivot_longer(cols=contains('cg'), names_to='CpG', values_to='Methylation')
notgap_samplecpg<-boxplotcpg%>%filter(!(CpG %in% row.names(mygaps$proberesults)))%>%pull(CpG)%>%unique()%>%head(n=4)

gap_plots<- ggplot(boxplotcpg%>%filter(CpG %in% row.names(mygaps$proberesults)), aes(x=childteen, y=Methylation, color=smkPreg_binary))+geom_jitter(position=position_dodge2(width=0.9), alpha=0.5)+geom_boxplot(alpha=0.8)+facet_wrap(~CpG)+theme(text=element_text(size=16))+scale_color_discrete(name='Prenatal smoke exposure?')+scale_x_discrete('', labels=c('Age 9', 'Age 15'))+theme_classic()+theme(legend.position='bottom')
not_gaps_plots<-ggplot(boxplotcpg%>%filter(!(CpG %in% row.names(mygaps$proberesults))), aes(x=childteen, y=Methylation, color=smkPreg_binary))+geom_boxplot()+facet_wrap(~CpG)+stat_compare_means(label='p.signif', label.y=0.9)+theme(text=element_text(size=16))+scale_color_discrete(name='Prenatal smoke exposure?')+theme(legend.position='bottom')+scale_x_discrete('', labels=c('Age 9', 'Age 15'))
not_gaps_sample<-ggplot(boxplotcpg%>%filter(CpG %in% notgap_samplecpg), aes(x=childteen, y=Methylation, color=smkPreg_binary))+geom_jitter(position=position_dodge2(width=0.9), alpha=0.5)+geom_boxplot(alpha=0.8)+facet_wrap(~CpG)+stat_compare_means(label='p.signif', label.y=0.9)+theme(text=element_text(size=16))+scale_color_discrete(name='Prenatal smoke exposure?')+scale_x_discrete('', labels=c('Age 9', 'Age 15'))+theme_classic()+theme(legend.position='bottom')
```


```{r save gap boxplots}
gap_plots

png(file.path(outputdir, 'GapPlots.png'), width=800, height=800)
gap_plots+theme(text=element_text(size=16))
dev.off()

not_gaps_plots

png(file.path(outputdir, 'NotGapPlotsSample.png'), width=800, height=800)
not_gaps_sample+theme(text=element_text(size=16))
dev.off()
```

```{r}
#read in top CpGs from Table One of Wicklund et al 2019
topCPGs<-read.csv(paste0(data.dir, "TopCpGsWiklundEtAl2019.csv"))
topCPGs<-topCPGs %>% mutate(estimate=as.numeric(gsub("−\\s", "-", gsub(" .*", "", β..SE.))), std.error=as.numeric(sub("\\).*", "", sub(".*\\(", "", β..SE.))), p.value=P.value, source="Wiklund et al. 2019", conf.low=(estimate-1.96*std.error), conf.high=(estimate+1.96*std.error), set ='')%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)

#read in Joubert meta top cpgs 
#read in cpgs from PMC4833289
cpgs<-read.csv(paste0(data.dir, 'PMC4833289_SuppFile2CpGs.csv'), stringsAsFactors = F, header=F, skip=2)
#make headers
header1<-scan(paste0(data.dir, 'PMC4833289_SuppFile2CpGs.csv'), nlines=1, what=character(), sep=',')
header1[7:26]<-c(rep("SSnewborn", 5), rep("SSnewbornCT", 5), rep("SSolder", 5), rep("anynewborn", 5))
header2<-paste(header1, scan(paste0(data.dir, 'PMC4833289_SuppFile2CpGs.csv'), nlines=1, skip=1, what=character(), sep=','), sep='_')
names(cpgs)<-header2

#four meta-EWASes: SS_newborn (sustained smoking in newborns); SS_newbornCT (sustained smoking in newborns controlling for cell type) SS_older (sustained smoking in older children); any_newborn (any smoking in newborns)
#pivot_longer to get single column of coefs and a column for the analysis type
#and group into a list by analysis type
cpgs<-cpgs %>% pivot_longer(cols = c(matches("_Coef|_SE|_P|_FDR|_Direction")), names_to = c("set", ".value"), names_pattern="(.+)_(.+)")%>%rename_with(~gsub('_', '', .x))%>%mutate(Gene=`Mapped Gene`, estimate=Coef, conf.low=estimate-1.96*SE, conf.high=estimate+1.96*SE, source='Joubert et al. 2016', std.error=SE, p.value=P, Chr=CHR) %>%mutate(source=paste0(source, ': ', set))%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)%>%filter(set %in% c('SSolder', 'SSnewbornCT'))


myCPGs<-AllModelsdf %>%filter(Model=='model1')%>%mutate(data=map(data, function(x) x %>%filter(CpG %in% sigCpGs)))%>%unnest(cols=c(data))%>%mutate(Joubertetal=ifelse(CpG %in% cpgs$CpG, 'in Joubert et al. 2016', 'Not in Joubert et al. 2016'), Wicklundetal=ifelse(CpG %in% topCPGs$CpG, 'in Wicklund et al. 2019', 'Not in Wicklund et al. 2019'))%>%mutate(Gene=gene, estimate=logFC, conf.low=CI.L, conf.high=CI.R, source=paste0('Fragile Families:', Age), std.error='NA', p.value=P.Value, Chr=chr, set='Model1', Position=pos)

#myCPGs<-child.m1.ss.hits.annotated %>% filter(adj.P.Val<0.2)%>%mutate(CpG=row.names(.))%>%mutate(Joubertetal=ifelse(CpG %in% cpgs$CpG, 'in Joubert et al. 2016', 'Not in Joubert et al. 2016'), Wicklundetal=ifelse(CpG %in% topCPGs$CpG, 'in Wicklund et al. 2019', 'Not in Wicklund et al. 2019'))%>%mutate(Gene=gene, estimate=logFC, conf.low=CI.L, conf.high=CI.R, source='Fragile Familes Age 9', std.error='NA', p.value=P.Value, Chr=chr, set='Model1', Position=pos)


cpg_overlap=myCPGs%>%filter(Joubertetal=='in Joubert et al. 2016' | Wicklundetal=='in Wicklund et al. 2019')%>%pull(CpG)%>%unique()

myCPGs2<-myCPGs %>%ungroup()%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)%>%mutate(CpG=ifelse(CpG %in% row.names(mygaps$proberesults), paste0(CpG, '*'), CpG))

cpg_all<-rbind(myCPGs2, cpgs, topCPGs)%>% filter(CpG%in% cpg_overlap)
cpgs_new<-myCPGs2%>%filter(!(CpG %in% cpg_overlap))

#geom blank to give good axes limits
blank_data <- cpg_all %>% filter(CpG%in% cpg_overlap)%>%dplyr::select(Gene, CpG, estimate, std.error)%>%group_by(Gene)%>%mutate(min=min(estimate), max=max(as.numeric(std.error), na.rm=T))%>%dplyr::select(-estimate, -std.error)%>%unique()%>%mutate(estimate=min*-1+max, conf.high=estimate, conf.low=estimate)%>%dplyr::select(-max, -min)%>%arrange(Gene)%>%mutate(source='Wiklund et al. 2019')


blank_data2 <- cpgs_new %>%dplyr::select(Gene, CpG, estimate, std.error)%>%group_by(Gene)%>%mutate(min=min(estimate))%>%dplyr::select(-estimate, -std.error)%>%unique()%>%mutate(estimate=min*-1+0.005, conf.high=estimate, conf.low=estimate)%>%dplyr::select(-min)%>%arrange(Gene)%>%mutate(source='Fragile Families:child')

plot_sharedCpG<-ggplot(cpg_all, aes(x=estimate, xmin=conf.low, xmax=conf.high, color=source, y=CpG))+geom_point(position=position_dodge(width=1))+geom_errorbarh(height=0.1, position=position_dodge(width=1))+scale_x_continuous()+geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+facet_wrap(.~Gene, scales='free', ncol=2)+theme_classic()+theme(legend.position = 'bottom')+guides(color=guide_legend(nrow=2,byrow=TRUE))+geom_blank(data = blank_data, aes(x = estimate))

plot_newCpG<-ggplot(cpgs_new, aes(x=estimate, xmin=conf.low, xmax=conf.high, color=source, y=CpG))+geom_point(position=position_dodge(width=1))+geom_errorbarh(height=0.1, position=position_dodge(width=1))+scale_x_continuous()+geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+facet_wrap(.~Gene, scales='free')+theme_classic()+theme(legend.position = 'bottom')+guides(color=guide_legend(nrow=2,byrow=TRUE))+geom_blank(data = blank_data2, aes(x = estimate))

cut_cpgs<-c(paste0(row.names(mygaps$proberesults), '*'), 'cg14055351')
plot_newCpG_cut<-ggplot(cpgs_new%>%filter(!(CpG%in%cut_cpgs)), aes(x=estimate, xmin=conf.low, xmax=conf.high, color=source, y=CpG))+geom_point(position=position_dodge(width=1))+geom_errorbarh(height=0.1, position=position_dodge(width=1))+scale_x_continuous()+geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+facet_wrap(.~Gene, scales='free')+theme_classic()+theme(legend.position = 'bottom')+guides(color=guide_legend(nrow=2,byrow=TRUE))+geom_blank(data = blank_data2%>%filter(!(CpG %in%cut_cpgs)), aes(x = estimate))
```



Filtering to only CpGs  that had an adjusted p-value of <0.2 in either the base child & teen models ($m_{CpG_{p_{BH}<0.2}}=$ `r length(sigCpGs)`), and then to only CpGs that overlapped with results from either Wicklund et al. 2019 or Joubert et al. 2016 ($m_{CpG_{overlap}}=$ `r length(cpg_overlap)` ), plot the estimates with error bars and compare. 

```{r, fig.width=10, fig.height=10}
plot_sharedCpG
png(file.path(outputdir, 'SharedCpG.png'), width=1000, height=800)
plot_sharedCpG+theme(text=element_text(size=18))
dev.off()
```

Also several CpGs that had an adjusted p-value <0.2 but were not overlapping with results from either Wicklund et al. 2019 or Joubert et al. 2016 ($m_{CpG_{notOverlap}}=$ `r cpgs_new$CpG %>%unique()%>%length()`). Some of these CpGs annotate to interesting immune (Il1A, CFB)  and neurology (ADNP, CLN8, GDNF, KIRREL2, RGNEF, SHANK2) genes. 

```{r, fig.width=10, fig.height=10}
plot_newCpG
png(file.path(outputdir, 'NewCpG.png'), width=1300, height=1000)
plot_newCpG+theme(text=element_text(size=18))
dev.off()
```
```{r, fig.width=10, fig.height=10}
plot_newCpG
png(file.path(outputdir, 'NewCpGCut.png'), width=1300, height=1000)
plot_newCpG_cut+theme(text=element_text(size=18))
dev.off()
```

# Longitudinal EWAS

```{r beta filtering and checks longitudinal, eval=F}
#load in betaqc matrix (trimmed)
#betaqc<-readRDS(file=paste0(data.dir, 'CreatedData/', 'betaTrim.Rds'))
#filter betaqc to modeldata samples only
beta_all <-betaqc[, colnames(betaqc) %in% modeldata$MethID]#; rm(betaqc)
#make sure that the order of the samples matches in the metadata matrix and the beta matrix are ordered the same
modeldata<-modeldata[match(colnames(beta_all), modeldata$MethID), ]
#checks
if(dim(beta_all)[2]!=dim(modeldata)[1]){print('STOP: Sample #s do not match')}
if(identical(colnames(beta_all), modeldata$MethID)==FALSE){print('STOP: Sample names do not match')}
```

```{r}

model_lmm<-function(df, y, x){
  xnew=paste0(x, '+', '(1|id)')
  lmer(formula(paste(y, xnew)), data=df, REML=F)
}

model_lmm<-function(y){
suppressWarnings(lmer(y~factor(smkPreg_binary)+global_PC1+global_PC2+factor(cm1bsex)+cm1inpov+ChildAgeComposite+Epi+IC+factor(smkPreg_binary)*ChildAgeComposite+(1|id), data=modeldata, REML=F))
}


data.frame(test_beta)%>%map(model_lmm)%>%map(tidy)%>%bind_rows(.id = 'cpg')

rm(betaqc)
bt=data.frame(t(beta_all)); rm(beta_all)
bt_test=bt[, 1:1000]

future::plan(multisession)

start.time=Sys.time()
mixed_models=bt_test%>%future_map(model_lmm, .progress=TRUE)%>%map(tidy)%>%bind_rows(.id = 'cpg')
end.time=Sys.time()
end.time-start.time

start.time=Sys.time()
mixed_models=bt_test%>%map(model_lmm)%>%map(tidy)%>%bind_rows(.id = 'cpg')
end.time=Sys.time()
end.time-start.time


```

```{r set models, eval=F}
mod1=model.matrix(~factor(modeldata$smkPreg_binary)+modeldata$global_PC1+modeldata$global_PC2+factor(modeldata$cm1bsex)+modeldata$cm1inpov+modeldata$ChildAgeComposite+modeldata$Epi+modeldata$IC+factor(modeldata$smkPreg_binary)*modeldata$ChildAgeComposite)

mod2=model.matrix(~factor(modeldata$smkPreg_binary)+modeldata$global_PC1+modeldata$global_PC2+factor(modeldata$cm1bsex)+modeldata$cm1inpov+modeldata$ChildAgeComposite+modeldata$Epi+modeldata$IC+factor(modeldata$m1g2_YesNoPreg)+factor(modeldata$m1g3_YesNoPreg)+factor(modeldata$smkPreg_binary)*modeldata$ChildAgeComposite)

mod3=model.matrix(~factor(modeldata$smkPreg_binary)+modeldata$global_PC1+modeldata$global_PC2+factor(modeldata$cm1bsex)+modeldata$cm1inpov+modeldata$ChildAgeComposite+modeldata$Epi+modeldata$IC+factor(modeldata$m1g2_YesNoPreg)+factor(modeldata$m1g3_YesNoPreg)+factor(modeldata$PostnatalMaternalSmokingAny)+factor(modeldata$SmkAtVisitPastmonth)+factor(modeldata$smkPreg_binary)*modeldata$ChildAgeComposite)

modellist=list(mod1, mod2, mod3)
modelnames=c('limmaModel1Long.Rda', 'limmaModel2Long.Rda', 'limmaModel3Long.Rda')
tophitnames_smk=c('LongModel1TopHits_smk.Rda', 'LongTeenModel2TopHits_smk.Rda', 'LongModel3TopHits_smk.Rda')
tophitnames_age=c('LongModel1TopHits_age.Rda', 'LongTeenModel2TopHits_age.Rda', 'LongModel3TopHits_age.Rda')
tophitnames_int=c('LongModel1TopHits_int.Rda', 'LongTeenModel2TopHits_int.Rda', 'LongModel3TopHits_int.Rda')
tophitnames=cbind(tophitnames_smk, tophitnames_age, tophitnames_int)
coef=c(2, 7, 10)
```

```{r loop through models longitudinal, eval=F}
#concern: For this function (duplicateCorrelation) to return statistically useful results, there must be at least two more arrays than the number of coefficients to be estimated, i.e., two more than the column rank of design.(from: https://rdrr.io/bioc/limma/man/dupcor.html)... because the design matrix is pretty large (many coefficients) is this a problem for us? 
#this is also taking forever, I think I should run the duplicateCorrelation on a subset of rows (see: https://support.bioconductor.org/p/117817/), see if it's normally distributed and then just use the median for everyone 
  print(paste('Fitting correlation structure for Model', i))
  corfit<-duplicateCorrelation(beta_all, design=modellist[[i]], block=modeldata$id)


for(i in 1:length(modellist)){
  print(paste('Fitting correlation structure for Model', i))
  corfit<-duplicateCorrelation(beta_all, design=modellist[[i]], block=modeldata$id)
  print(paste('Running Model', i))
  out=lmFit(beta_all, modellist[[i]], block=modeldata$id, correlation=corfit$consensus)
  out=eBayes(out)
  saveRDS(out, file=file.path(limma.out.dir, modelnames[i]))
  for(j in 1:3){
    print(paste('Annotating top hits from model', i, 'coef', coef[j]))
    long.ss.hits<-topTable(out, coef=coef[j], number=nrow(beta_all), confint=T)
    long.ss.annotated.hits<-add_annotations(long.ss.hits)
    saveRDS(long.ss.annotated.hits, file=file.path(limma.out.dir, tophitnames[i, j]))
    rm(long.ss.hits, long.ss.annotated.hits)
  }
  rm(out)
}
```
