---
title: "Supplemental Materials: Polymethylation Scores for Prenatal Maternal Smoke Exposure Persist Until Age 15 and Are Detected in Saliva"
author: "Freida A. Blostein^1, Jonah Fisher^2, John Dou^1, Lisa Schenper^3, Erin Ware^2 Daniel A. Notterman^3, Colter Mitchell^2, Kelly M. Bakulski^1"
date: "`9/30/2021`r Sys.Date()`"
output:
  html_document:
    code_folding: show
    highlight: tango
    number_sections: no
    theme: sandstone
  word_document: default
  pdf_document: default
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message=FALSE)

library(ggplot2)
library(kableExtra)
library(ggpubr)
library(tidyr)
library(stringr)
library(dplyr)
library(compareGroups)
library(kableExtra)
library(gtsummary)
library(gtools)
library(haven)
library(sjlabelled)
library(matchmaker)
library(EpiDISH)
library("PerformanceAnalytics")
library(DescTools)
library(purrr)
library(broom)
library(broom.mixed)
library(ggplot2)
library(multcomp)
library(ggpubr)
library(cowplot)
library(geepack)
library(lme4)
library(pROC)
library(lmerTest)
library(stringr)
library(sva)
library(gtsummary)
library(gt)
library(here)
library(DiagrammeRsvg)
library(xml2)
library(ggrepel)

as_supp=F
stand_alone=F
load_data=T
run_scripts=F
```

```{r temp read n data, eval=load_data}
basedir=here::here()
outdir=file.path(gsub('Code', 'Output', basedir), 'Figures')
datadir=gsub('Code', 'Data', basedir)
scriptdir=file.path(basedir, 'Scripts')
load(paste0(datadir, '/CreatedData/allPhenoData.Rdata'))
load(paste0(datadir, '/CreatedData/allmethyl.Rdata'))
load(paste0(datadir, '/CreatedData/completeCaseSVA.Rdata'))
load(paste0(datadir, '/CreatedData/methylation_summarymodels.Rdata'))
modeldata=modeldata %>% mutate(childteen=factor(childteen, levels=c('Age 9', 'Age 15')))%>%droplevels()%>%copy_labels(modeldata)
```

```{r sources scripts instead of loading, eval=F}
source(file.path(scriptdir, '01_clean_phenotype_data.R'))
source(file.path(scriptdir, '02_make_methyl_summary.R'))
#assuming 03_sva_variables.R has been run and exists separately as this would take a day to run. 
source(file.path(scriptdir, '04_summary_methyl_models.R'))
```

```{r plotting variables}
theme_set(theme_bw())
set_label(modeldata$cm1bsex)='Child gender'
set_label(modeldata$cm1inpov)='Maternal income to poverty threshold (at baseline)'
set_label(modeldata$PostnatalMaternalSmokingDose)='Primary caregiver smoking past month by dose'
set_label(modeldata$SSnewbornCT_center)='Polymethylation score for smoke exposure'
set_label(modeldata$PedBE)='Pediatric epigenetic clock (years)'
set_label(modeldata$DNAmGrimAge)='GRIM epigenetic clock (years)'
set_label(modeldata$globalmethylation)='Percent global DNA methylation'
set_label(modeldata$cg05575921)='AHRR gene: \n percent cg05575921 methylation'
set_label(modeldata$PostnatalMaternalSmokingAny)='Postnatal maternal smoking at ages 1 or 5'
set_label(modeldata$SmkAtVisitPastmonth)='Postnatal primary caregiver smoking in past month prior to visit'
set_label(modeldata$childteen)='Visit of DNA methylation measure'
```

```{r set options and record n}
n_total=modeldata %>% dplyr::select(id)%>%unique()%>%nrow()
m_obs=modeldata%>%nrow()
n_twovisit=modeldata %>% group_by(id)%>%summarise(n=n())%>%filter(n==2)%>%nrow()

all_methyl=c("globalmethylation", "PedBE", "DNAmGrimAge", 'DNAmPACKYRS', "Epithelial.cells_saliva", "Leukocytes_saliva", "SSnewbornCT_center", "richmond568_center", "richmond19_center", "reese_center", "enscore_probs", "cg05575921")
main_methyl=c("globalmethylation", "PedBE", "SSnewbornCT_center", "cg05575921")


all_methyl_labels=modeldata%>%dplyr::select(any_of(all_methyl))%>%get_label()
#all_methyl_labels[str_detect(all_methyl_labels, 'Polymethylation', )]=paste0(rep('Score:', 3), c('Sustained smoking \n newborns', 'Sustained smoking \n newborns (w/ CT control)', 'Sustained smoking \n older children'))
all_methyl_labels[str_detect(all_methyl_labels, 'Polymethylation score for smoke exposure')]='6074 site polymethylation score: coefficients for sustained smoking from Joubert newborn cord blood meta-analysis with cell-type correction (mean-centered)'
names(all_methyl_labels)=all_methyl

all_methyl_levels=c("SSnewbornCT_center", 'richmond568_center', 'richmond19_center', 
                    'reese_center', 'enscore_probs', "cg05575921", 
                     "PedBE" , "DNAmGrimAge", 'DNAmPACKYRS', 
                     "globalmethylation","Epithelial.cells_saliva", "Leukocytes_saliva")

main_methyl_labels=modeldata%>%dplyr::select(any_of(main_methyl))%>%get_label()


my_round=function(x, d){
  y=c()
  d_orig=d
  for(i in 1:length(x)){
    d=d_orig
    while(round(x[i], d)==0){
    d=d+1
    }
    y[i]=round(x[i], d)
  }
  return(y)
}
```

```{r figure table counter}
main_fig=1
sup_fig=1
main_tab=1
sup_tab=1
```


```{r figure 1, eval=T}
#function to make a pretty count of n and m 
N_M_count=function(filter_statement, statement, df){
  paste0('m=', df %>% filter(eval(parse(text = filter_statement)))%>%nrow(), 
         ' samples & ', 
         'n=', df %>% filter(eval(parse(text = filter_statement))) %>% dplyr::select(idnum) %>%unique()%>%nrow(), 
         ' individuals with', statement)
}
#Make Labels
n_fullFF<-paste0("Fragile Families cohort \n", "n=", FF_labeled %>% nrow())

#original number of samples and individuals with methylation data
all_samples=readRDS(file.path(datadir, 'OGData', 'pd_qc.rds'))
all_samples_nomoms=all_samples %>%filter(childteen!='M')
m_og_methyl=nrow(all_samples_nomoms) # number of samples - 12 mother samples
n_og_methyl=length(unique(all_samples_nomoms$idnum)) # number of individuals 
n_methylQC<-glue::glue("Cohort with 450K methylation data \n  n={n_og_methyl} individuals with m={m_og_methyl} samples")

#These numbers come from Jonahs preprocessing document, I save them as variables to ensure correspondence between Figure 1 and the document text
probe_fail=34; sex_mismatch=11; meth_outlier=6; genetic_discordance=27; tech_rep=49;idat_irregularity=1

n_QCloss<-glue::glue("Methylation data QC: ", 
                 '{idat_irregularity} sample with idat file irregularity
                 {probe_fail} samples with probe failure percentage>10%
                {sex_mismatch} samples with predicted sex mismatch 
                 {meth_outlier} samples with outlier methylation values
                 {genetic_discordance} samples with genetic discordance
                 {tech_rep} technical replicates')

n_methylFF<-paste0("Cohort with quality controlled 450K data \n","n=", myFF %>% dplyr::select(id) %>% unique() %>% nrow(), " individuals with m=", myFF%>%nrow(), " samples")


n_missingmaternalprenataldata=paste0('Missing maternal data: \n', 
                                     N_M_count("is.na(smkPreg_binary)", " with missing prenatal maternal smoking \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & is.na(m1g2_YesNoPreg)", " with missing prenatal alcohol \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & is.na(m1g3_YesNoPreg)", 
                                               " with missing prenatal other drug use \n ", myFF),
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & is.na(cm1inpov)", 
                                               " with missing maternal income to poverty ratio data \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) &
                                               PostnatalMaternalSmokingAny=='Missing'", " with missing early life postnatal smoke exposure data \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) &
                                               PostnatalMaternalSmokingAny!='Missing' & SmkAtVisitPastmonth=='Missing'", 
                                               " with missing past month primary care giver smoking packs/day", myFF))

maternaldata=myFF %>% filter(!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) & PostnatalMaternalSmokingAny!='Missing' & SmkAtVisitPastmonth!='Missing')  

n_withmaternaldata=paste0('m=', maternaldata %>% nrow(), ' samples & n=', maternaldata %>% dplyr::select(idnum)%>%unique()%>%nrow(), ' individuals with maternal covariate data')
  
n_missingchilddata=paste0('Missing child data: \n', 
                          N_M_count("is.na(ancestry) | ancestry=='Missing PC data'", 
                                    " with missing ancestry/genetic data  \n ", maternaldata), 
                          N_M_count("!is.na(ancestry) & ancestry!='Missing PC data' & is.na(cm1bsex)", 
                                    " with missing child sex data \n ", maternaldata), 
                          N_M_count("!is.na(ancestry) & ancestry!='Missing PC data' & !is.na(cm1bsex) & is.na(ChildAgeComposite)", 
                                    " with missing child age at sample \n ", maternaldata))

childdata= maternaldata %>% filter(ancestry!='Missing PC data' & !is.na(cm1bsex) & !is.na(ChildAgeComposite))
n_withchilddata=paste0('m=', childdata %>% nrow(), ' samples & n=', childdata %>% dplyr::select(idnum)%>%unique()%>%nrow(), ' individuals with child covariate data')

n_Child<-paste0('n=', childdata %>%filter(childteen=='C')  %>% nrow(), " individuals with a 9 year visit")
n_ChildSmkexcluded<-paste0("Exclude children who smoke: \n",'n=', childdata%>%filter((k5f1l=="2 no" | (is.na(k5f1l) & k6d40=='2 No')) & childteen=='C')%>% nrow(), 
                           " individuals with a 9 year visit")
n_ChildSmkexclusions<-paste0('n=', childdata %>% filter((k5f1l=="1 yes" | (is.na(k5f1l) & (k6d40!='2 No' | is.na(k6d40)))) & childteen=='C') %>% nrow(), 
                             " individuals smoking \n or missing child smoking data \n at 9 year visit")

n_Teen<-paste0('n=', childdata %>%filter(childteen=='T')   %>% nrow(), " individuals with a 15 year visit")
n_TeenSmkexcluded<-paste0("Exclude children who smoke: \n",'n=', childdata %>%filter(childteen=='T' & k6d40=="2 No" & (k5f1l!='1 yes' | is.na(k5f1l))) %>% nrow(), 
                          " individuals with a 15 year visit")
n_TeenSmkexclusions<-paste0('n=', childdata %>% filter((k6d40=="1 Yes" | is.na(k6d40) | k5f1l=='1 yes') & childteen=='T') %>% nrow(),
                            " individuals smoking \n or missing focal child smoking data \n at 15 year visit") 


test=childdata %>%filter((childteen=='T' & k6d40=="2 No") | (childteen=='C' & k5f1l=='2 no'))

#if(childdata %>% filter(((k5f1l=="2 no" | (is.na(k5f1l) & k6d40=='2 No')) & childteen=='C') | (childteen=='T' & k6d40=="2 No" & (k5f1l=='2 no' |  is.na(k5f1l))))%>%nrow()!=nrow(modeldata)){stop('filter N and completecase data set N do not match')}


```

```{r figure 1 make figure, fig.height=12}

graph1<-DiagrammeR::grViz("
   digraph graph2 {
   compound=true;
   graph [layout = dot]
   
   node [shape = rectangle, width=4]
   a [label = '@@1']
   b [label = '@@2']
   c [label = '@@3']
   d [label = '@@4']
   e [label = '@@5']
   f [label = '@@6']
   g [label = '@@7']
   h [label = '@@11']
   i [label = '@@12']
   

   #fake nodes for connecting to edges
   node [shape=point, width=0.01, height=0.01]
   a1
   b1
   c1
   d1
   f1 [group=g1]
   g1
   h1
  
  #exclusion nodes
  node [shape = rectangle, width=4]
  aa [label = '@@8']
  bb [label = '@@9']
  cc [label = '@@10']
  hh [label = '@@13']
  ii [label = '@@14']
    
  #draw connections
#  a->m
#  m->a1 [arrowhead=none]
  a->a1 [arrowhead=none]
  a1->b
  b->b1 [arrowhead=none]
  b1->c
  c->c1 [arrowhead=none]
  c1->d
  d->d1 [arrowhead=none]
  d1->e
  e->f; e->g
  f->g1 [arrowhead=none]; g->h1 [arrowhead=none]
  g1->h; h1->i 
  
  
{rank=same; b1->aa [minlen=2]}   
{rank=same; c1->bb [minlen=2]} 
{rank=same; d1->cc [minlen=2]}
{rank=same; g1->hh [minlen=2]}
{rank=same; h1->ii [minlen=2]}



#invisible constraints to force ordering
#ff->i [style=invis]
   }   
   
   [1]: n_fullFF
   [2]: n_methylQC
   [3]: n_methylFF
   [4]: n_withmaternaldata
   [5]: n_withchilddata
   [6]: n_Child
   [7]: n_Teen
   [8]: n_QCloss
   [9]: n_missingmaternalprenataldata
   [10]: n_missingchilddata
   [11]: n_ChildSmkexcluded
   [12]: n_TeenSmkexcluded
   [13]: n_ChildSmkexclusions
   [14]: n_TeenSmkexclusions
")

graph1 %>%
  export_svg() %>%
  read_xml() %>%
  write_xml(file.path(outdir, paste0("Figure", main_fig, ".svg")))

save_png <- function(plot, path){
  DiagrammeRsvg::export_svg(plot) %>%
    charToRaw() %>%
    rsvg::rsvg() %>%
    png::writePNG(path)
}
save_png(graph1, file.path(outdir, paste0('Figure', main_fig, ".png")))

 DiagrammeRsvg::export_svg(graph1) %>%
    charToRaw() %>%
    rsvg::rsvg_ps(file.path(outdir, paste0('Figure', main_fig, ".ps")))

main_fig=main_fig+1            
```

```{r figure 1 plot figure, eval=stand_alone}
graph1
```

# Supplemental Table `r sup_tab` - Descriptive statistics of excluded vs included samples with methylation data available from the Fragile and Families and Child Wellbeing study. Samples excluded according to the flow diagram in Main Figure 1 

```{r supplemental table exclusions}
myFF$MethylData2<-ifelse(myFF$MethID %in% modeldata$MethID, 'Final analysis group', 'Not in final analysis group')
set_label(myFF$MethylData2)<-'Exclusions based on missing data'
myFF=left_join(myFF, methyldata)%>%mutate(idnum=as.character(idnum))
myFF=myFF%>%mutate(across(c(PostnatalMaternalSmokingAny, SmkAtVisitPastmonth), ~ifelse(.x=='Missing', 
                                                       NA, .x)))%>%
  mutate(ancestry=ifelse(ancestry=='Missing PC data', NA, ancestry))%>%copy_labels(myFF)

exclusion_tbl=myFF %>% mutate(childteen=case_when(childteen=='C'~'Age 9', childteen=='T'~'Age 15'))%>%
    dplyr::select(smkPreg_binary, childteen, MethylData2, SSnewbornCT_center,  richmond568_center, reese_center, enscore_probs, cg05575921, PedBE, globalmethylation, Leukocytes_saliva, Epithelial.cells_saliva, ancestry, cm1bsex, city_binary, k6d40, k5f1l, m1g2_YesNoPreg, m1g3_YesNoPreg, cm1inpov, PostnatalMaternalSmokingAny,SmkAtVisitPastmonth )%>%
  mutate(smkPreg_binary=factor(smkPreg_binary, levels=c('No', 'Yes'), labels=c('Not exposed', 'Smoke exposed')), 
         cm1bsex=factor(cm1bsex, levels=c('1 Boy', '2 Girl'), labels=c('Boy', 'Girl')))%>%
  mutate(cm1bsex=set_label(cm1bsex, 'Child gender'), smkPreg_binary=set_label(smkPreg_binary, 'Prenatal maternal smoking'),
         childteen=set_label(childteen, 'Age at DNA methylation sample'))%>%
  tbl_summary(., by=MethylData2, statistic=list(all_continuous()~"{mean} ({sd})"))%>%
  modify_header(stat_by = "**{level}**, M samples = {n}")%>%
  add_n(col_label='M samples')#%>%
  #add_p(test=list(all_continuous()~"t.test"))

gtsave(as_gt(exclusion_tbl), file.path(outdir, paste0('SupplementalTable', sup_tab, '.rtf')))

sup_tab=sup_tab+1
```

```{r exclusion table include, eval=as_supp}
exclusion_tbl
```

```{r wide data}
modeldata_wide = modeldata%>%
  mutate(childteen=gsub(' ', '_', childteen))%>%
  dplyr::select(any_of(c('idnum', 'ancestry', 'childteen', 'cm1bsex', 'ancestry', 
                         'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'cm1inpov',
                         'smkPreg_binary',  all_methyl)))%>%
  pivot_wider(names_from = childteen, values_from=any_of(all_methyl))

model_twovisit=modeldata_wide[complete.cases(modeldata_wide), ]

modeldata_people=modeldata%>%
  dplyr::select(any_of(c('idnum', 'ancestry', 'cm1bsex', 'ancestry', 
                         'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'cm1inpov',
                         'smkPreg_binary', 'PostnatalMaternalSmokingAny', 'city_binary')))%>%unique()
```


# Supplemental Figure `r sup_fig` - Pearson correlation of selected DNA methylation summary measures from an anlysis of M=`r m_obs` samples and N=`r n_total`children in the Fragile Families and Child Wellbeing cohort. Lower diaganol shows the Pearson rho correlation coefficient, upper triangle shows ellipses illustrating relationship between variables colored by the magnitude and direction of the correlation. 

```{r make supplemental figure correlation plot}
library(Hmisc)
library(corrplot)
res2<-modeldata%>%dplyr::select(any_of(all_methyl))%>%cor()
#labels
colnames(res2)=str_wrap(all_methyl_labels, width=70)
rownames(res2)=str_wrap(all_methyl_labels, width=70)

pdf(file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=12)
corrplot.mixed(res2, upper="ellipse", lower='number', order='hclust', tl.pos ="lt", tl.cex=0.5, number.cex= 7/ncol(res2))
dev.off()

sup_fig=sup_fig+1
```

```{r supplemental correlation plot plot it , eval=as_supp, fig.width=12, fig.height=12}
corrplot.mixed(res2, upper="ellipse", lower='number', order='hclust', tl.pos ="lt", tl.cex=0.8, number.cex= 7/ncol(res2))

```

# Supplemental Figure `r sup_fig` - Correlation between age 9 and age 15 visit DNA methylation summary measures among `r n_twovisit` children with age 9 and age 15 visit data in the Fragile Families and Child Wellbeing study. Each point represents one individual, x-axis is the DNA methylation summary measure at age 9 and y-axis is the DNA methylation measure at age 15. Blue lines are simple linear regression lines of best fit, with Pearson rho and associated $P$ value shown. From top-left: Polymethylation score for prenatal maternal smoking, score created using coefficients from a regression of prenatal maternal smoking and DNA methylation in newborn cord blood, incorporating cell-type control (Joubert et al. 6074 DNA methylation sites). Polymethylation score for prenatal maternal smoking using coefficients from a regression of prenatal maternal smoking and DNA methylation in newborn cord blood, no cell-type control (Joubert et al. 568 DNA methylation sites). Polymethylation score calculated using regression coefficients from a regression of prenatal maternal smoking and DNA methylation in older children's peripheral blood samples (Joubert et al. 19 DNA methylation sites). Polymethylation score calculated using regression coefficients from a LASSO regression of prenatal maternal smoking and DNA methylation in newborn cord blood samples (Reese et al. 28 DNA methylation sites). Polymethylation score calculated using coefficients from an elastic net regression of prenatal maternal smoking and DNA methylation in adolescents peripheral blood (Rauschert et al. 204 site score). Percent DNA methylation at cg05575921 in AHRR.  Pediatric epigenetic clock (years). GRIM epigenetic clock (years). GRIM epigenetic clock pack years component. Percent global DNA methylation (mean across all DNA methylation sites in cleaned probe set). Proportion of epithelial cells calculated from DNA methylation using reference panel of children's saliva. Proportion of immune cells calculated from DNA methylation using a reference panel of children's saliva. 

```{r supplemental figure time correlation plot}
age_long_methyl=modeldata%>%
  dplyr::select(any_of(c('idnum', 'childteen', 'ancestry', 'smkPreg_binary', all_methyl)))%>%
  pivot_longer(cols=any_of(all_methyl),
               names_to='Methylation summary measure type',
               values_to='Methylation summary measure')%>%
  mutate(Methyl_factor=factor(`Methylation summary measure type`, 
                              levels=all_methyl_levels))

age_wide_methyl=age_long_methyl%>%
  pivot_wider(names_from=childteen, 
              values_from=`Methylation summary measure`)

#checks
if(age_wide_methyl%>%filter(Methyl_factor=='globalmethylation')%>%filter(!is.na(`Age 9`) & !is.na(`Age 15`))%>%nrow()!=n_twovisit){stop('check sample size two visit vs age_wide methyl')}


visit_cor=age_wide_methyl%>%
  ggplot(aes(x=`Age 9`, y=`Age 15`))+geom_point(size=1)+stat_cor(label.sep=', ', size=2.5, label.y.npc=0.95)+geom_smooth(method=lm)+theme_classic()+
  scale_x_continuous('Age 9 score')+scale_y_continuous('Age 15 score')+
  facet_wrap(~Methyl_factor, scales='free', labeller = as_labeller(all_methyl_labels, default=label_wrap_gen(60)), ncol=3)+
  theme(strip.text = element_text(size=5))

ggexport(visit_cor, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1
```

```{r supplemenetal figure time correlation plot, eval=as_supp, fig.width=12, fig.height=12}
visit_cor
```

# Supplemental Figure `r sup_fig`: Distributions of selected methylation summary measures at age 9 and age 15 among `r n_total` children in the Fragile Families and Child Wellbeing study. Captive box plots within violin plots show the distribution of selected DNA methylation measures in samples from age 9 (red) and age 15 (blue). Black lines show lines of best fit. From top-left: Polymethylation score for prenatal maternal smoking, score created using coefficients from a regression of prenatal maternal smoking and DNA methylation in newborn cord blood, incorporating cell-type control. Polymethylation score for prenatal maternal smoking using coefficients from a regression of prenatal maternal smoking and DNA methylation in newborn cord blood, no cell-type control. Polymethylation score calculated using regression coefficients from a regression of prenatal maternal smoking and DNA methylation in older children's peripheral blood samples. Pediatric epigenetic clock (years). GRIM epigenetic clock (years). Percent DNA methylation at cg05575921 in AHRR. Percent global DNA methylation (mean across all DNA methylation sites in cleaned probe set). Proportion of epithelial cells calculated from DNA methylation using reference panel of children's saliva. Proportion of immune cells calculated from DNA methylation using a reference panel of children's saliva. 

```{r supplemental figure methylation time distributions}
dist_methyl=age_long_methyl%>%
  ggplot(aes(x=childteen, y=`Methylation summary measure`, color=childteen))+
  geom_violin()+geom_boxplot(width=0.2)+stat_compare_means(label="..p.signif..", label.y.npc=0.87)+
  scale_color_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  scale_x_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  geom_point(alpha=0)+geom_line(alpha=0, aes(group=idnum))+geom_smooth(aes(group=1), method='lm', color='black')+
  facet_wrap(~Methyl_factor, scales='free', labeller = as_labeller(all_methyl_labels, default=label_wrap_gen(60)), ncol=3)+
  theme(legend.position = 'bottom')+
  theme(strip.text = element_text(size=5))


ggexport(dist_methyl, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1

```

```{r supplemental figure methylation time distribution, eval=as_supp, fig.height=11}
dist_methyl
```


```{r main figure 2 change in scores by smoking status, fig.width=12}

#add line for mean ? 
outcome_labels=c('Global methylation', 'Pediatric clock', 'AHRR: cg05575921', '6074 site polymethylation score \n (newborn, mean centered)', 'Polymethylation score  \n (older children, mean centered')
names(outcome_labels)=c('globalmethylation', 'PedBE', 'cg05575921', 'SSnewbornCT_center','SSolder_center')

fig2=age_long_methyl%>%
    filter(Methyl_factor%in% main_methyl)%>%
    ggplot(aes(x=childteen, y=`Methylation summary measure`, color=smkPreg_binary))+
    geom_violin()+geom_boxplot(width=0.2, position = position_dodge(0.9))+
  stat_compare_means(label='..p.signif..', label.y.npc = 0.88)+
    facet_wrap(~Methyl_factor, scales='free',labeller = as_labeller(all_methyl_labels, default=label_wrap_gen(40)))+
    scale_x_discrete('Visit', labels=c('Age 9', 'Age 15'))+
    scale_color_grey('', labels=c('Not prenatal maternal smoke exposed', 'Prenatal smoke exposed'))+
    theme(legend.position = 'bottom')




ggexport(fig2, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')))
ggexport(fig2, filename=file.path(outdir, paste0('Figure', main_fig, '.eps')))
main_fig=main_fig+1

```

```{r main figure 2 change in scores smoking status, eval=stand_alone}
fig2
```


```{r bivariate table}

smoking_tbl=modeldata %>% 
  dplyr::select(smkPreg_binary, childteen, SSnewbornCT_center, cg05575921, PedBE, globalmethylation, Leukocytes_saliva, Epithelial.cells_saliva, ancestry, cm1bsex,  city_binary, m1g2_YesNoPreg, m1g3_YesNoPreg, cm1inpov, PostnatalMaternalSmokingAny,SmkAtVisitPastmonth )%>%
  mutate(smkPreg_binary=factor(smkPreg_binary, levels=c('No', 'Yes'), labels=c('Not exposed', 'Smoke exposed')), 
         cm1bsex=factor(cm1bsex, levels=c('1 Boy', '2 Girl'), labels=c('Boy', 'Girl')))%>%
    mutate(cm1bsex=set_label(cm1bsex, 'Child gender'), smkPreg_binary=set_label(smkPreg_binary, 'Prenatal maternal smoking'),
         childteen=set_label(childteen, 'Age at DNA methylation sample'))%>%
  tbl_strata(strata=childteen, .tbl_fun=~.x %>%
               tbl_summary(by=smkPreg_binary, statistic=
                             list(all_continuous()~"{mean} ({sd})"))%>%
               add_n()%>%add_p(test=list(all_continuous()~"t.test")))


smoking_tbl=smoking_tbl%>%as_gt()%>%
  tab_row_group(label='Child characteristics', rows = 1:13)%>%
  tab_row_group(label='Maternal characteristics', rows=14:23)%>%
     row_group_order(
      groups = c("Child characteristics", "Maternal characteristics")
    )

gtsave(smoking_tbl, file.path(outdir, paste0('Table', main_tab, '.rtf')))


main_tab=main_tab+1
```

```{r, eval=stand_alone}
smoking_tbl
```

# Supplemental Table `r sup_tab`:  Bivariate associations between prenatal maternal smoking and additional DNA methylation summary measures among a diverse sample of `r n_total` children in the Fragile Families and Child Wellbeing study. Percent DNA methylation shown calculated across all DNA methylation sites available (global DNA methylation) and stratified by genomic region (Island, Open Sea, Shelf, Shore). Polymethylation score for smoke exposure from main analysis and from sensitivity analyses where scores were calculated using alternative available regression coefficients. 

```{r supplemental table additional methylation measures}
smoking_tbl_extendedMeth=modeldata %>% 
  dplyr::select(smkPreg_binary, globalmethylation, Island_mean, OpenSea_mean, Shelf_mean, Shore_mean, 
                PedBE, DNAmGrimAge, DNAmPACKYRS, Epithelial.cells_saliva, Leukocytes_saliva,
                SSnewbornCT_center, 
                richmond568_center, richmond19_center, reese_center, enscore_probs,
                matches('cg'),  childteen, -PCGrimAge)%>%
    mutate(smkPreg_binary=factor(smkPreg_binary, levels=c('No', 'Yes'), labels=c('Not exposed', 'Smoke exposed')))%>%
  tbl_strata(strata=childteen, .tbl_fun=~.x %>%
               tbl_summary(by=smkPreg_binary, statistic=
                             list(all_continuous()~"{mean} ({sd})"))%>%
               add_n()%>%add_p(test=list(all_continuous()~"t.test")))

gtsave(as_gt(smoking_tbl_extendedMeth), file.path(outdir, paste0('SupplementalTable', sup_tab, '.rtf')))

sup_tab=sup_tab+1

```

```{r supplemental table additional methylation measures print, eval=as_supp}
smoking_tbl_extendedMeth
```


```{r main figure 3 forest plot}
my_outcomes=c('Global methylation', 'Pediatric clock', 'AHRR: cg05575921', '6074 site polymethylation score: coefficients for sustained smoking from Joubert newborn cordblood meta-analysis, w/ cell-type control - PMC4833289 (/n mean-centered (coefficients) & z-score standardized (score))')

outcome_labels=c('Percent global  \n DNA methylation  (percent)', 'Pediatric epigenetic \n clock (years)', 'Percent DNA methylation of \n cg05575921 in AHRR (percent)',  '6074 site polymethylation score for prenatal smoke exposure with cell-type corrected coefficients (standard deviations)')

all_models=rbind(local_models, global_models%>%mutate(ancestry='Unstratified'))%>%
  mutate(childteen=factor(childteen, levels=c('Age 15', 'Age 9')))

data_text=all_models%>%
  mutate(n=map_dbl(data, nrow), n=paste0('n=', n), x=-0.5)%>%
  unnest(tidied)%>%
  filter(str_detect(outcome, 'cell-type control') & term=='smkPreg_binaryYes'& predictors=='base model')%>%mutate(outcome_f=outcome)
  
names(outcome_labels)=my_outcomes

forest_plot=all_models%>%
  mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
  unnest(tidied)%>%
  filter(term=='smkPreg_binaryYes' & predictors=='base model')%>%
  filter(outcome%in% my_outcomes)%>%
    mutate(outcome_f=factor(outcome, levels=rev(my_outcomes)))%>%
  ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, levels(age)), y=ancestry))+
  geom_point(position = position_dodge(width = 1))+
  geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
  facet_grid(.~outcome_f, scales='free_x', labeller = as_labeller(outcome_labels, default=label_wrap_gen(50)))+
  geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
  geom_text(data=data_text, aes(x=x,label=n), position=position_dodge(width = 1), hjust=0.05)+
  theme_classic()+
  theme(text = element_text(size=16))+theme(legend.position = 'bottom')+
  scale_x_continuous('Coefficient for prenatal maternal smoking + 95% CI')+
  scale_y_discrete('', limits=c('European ancestry', 'Admixed ancestry - Latin heritage', 'African ancestry', 'Unstratified'), labels=str_wrap(c('European genetic ancestry', 'Admixed genetic ancestry - Latin heritage', 'African genetic ancestry', 'Unstratified'), 30))+scale_color_manual('', values=c('black', 'darkgrey'), breaks=c('Age 9', 'Age 15'))


ggexport(forest_plot, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=19)
ggexport(forest_plot, filename=file.path(outdir, paste0('Figure', main_fig, '.eps')), width=19)
main_fig=main_fig+1
```

```{r main figure 3 print forest plot, eval=stand_alone, fig.width=12}
forest_plot
```

# Supplemental Figure `r sup_fig`: Sensitivity analyses for controlling for different variable sets in multivariable analyses of prenatal maternal smoking and DNA methylation summary measures. Base model: Child sex, maternal income to poverty ratio, immune cell proportion, plate, 1st two principal components from PCA on child genetic data, child age. Prenatal exposures model: base model + yes/no any prenatal maternal alcohol use, yes/no any prenatal other drug use. Secondhand smoking models: prenatal exposures model + yes/no any postnatal maternal/primary care giver smoking at age 1 or 5, pk/day in month prior to primary caregiver/maternal interview at age 9 or 15. Surrogate variable: including all surrogate variables


```{r supplemental figure all models forest}
#with(modeldata, table(ancestry, PostnatalMaternalSmokingAny, smkPreg_binary))%>%kbl()

forest_plot_compareModels=all_models%>%
    mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
    unnest(tidied)%>%
    filter(term=='smkPreg_binaryYes')%>%
    filter(outcome%in% my_outcomes)%>%
    mutate(outcome_f=factor(outcome, levels=rev(my_outcomes)))%>%
    ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, levels(age)), y=ancestry))+
    geom_point(position = position_dodge(width = 1))+
    geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
    facet_grid(predictors~outcome_f, scales='free_x', labeller = labeller(outcome_f=outcome_labels))+
    geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
    theme_classic()+
    theme(text = element_text(size=16))+theme(legend.position = 'bottom')+
    scale_x_continuous('Coefficient for prenatal maternal smoking + 95% CI')+
    scale_y_discrete('', limits=c('European ancestry', 'Admixed ancestry - Latin heritage', 'African ancestry', 'Unstratified'), labels=c('European genetic ancestry', 'Admixed ancestry - Latin heritage', 'African genetic ancestry', 'Unstratified'))+scale_color_manual('', breaks=c('Age 9', 'Age 15'), values=c('black', 'darkgrey'))

ggexport(forest_plot_compareModels, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=18, height=15)
sup_fig=sup_fig+1
```

```{r, eval=as_supp, fig.width=14, fig.height=12}
forest_plot_compareModels
```

# Supplemental Figure `r sup_fig` DNA methylation summary measures across combinations of exposure to prenatal and postnatal  maternal smoking. Sample size is shown at the bottom of violin plots with captured boxplots. POlymethylation score shown in top facets, percent DNA methylation at cg05575921 in AHRR gene shown in bottom facets, age 9 samples shown in left facets, age 15 shown in right facets

```{r supplemental figure pre post smoke exposure}

my_comparisons <- list(c("No prenatal smoke exposed & No maternal smoking at age 1 and 5", 
                        "No prenatal smoke exposed & Maternal smoking at age 1 or age 5"), 
                        c("No prenatal smoke exposed & Maternal smoking at age 1 or age 5", 
                          "Yes prenatal smoke exposed & Maternal smoking at age 1 or age 5"), 
                      c("Yes prenatal smoke exposed & No maternal smoking at age 1 and 5", 
                        "Yes prenatal smoke exposed & Maternal smoking at age 1 or age 5"))

n_count=modeldata%>%mutate(exposed=paste0(smkPreg_binary, ' prenatal smoke exposed & ', PostnatalMaternalSmokingAny))%>% group_by(childteen, exposed)%>%summarise(n=n())

give.n <- function(x){
  return(c(y = min(x)-0.1, label = length(x))) 
  # experiment with the multiplier to find the perfect position
}

dna_prepost=modeldata%>%
  dplyr::select(any_of(c('idnum', 'childteen', 'ancestry', 'smkPreg_binary', 'PostnatalMaternalSmokingAny', all_methyl)))%>%
  pivot_longer(cols=any_of(all_methyl),
               names_to='Methylation summary measure type',
               values_to='Methylation summary measure')%>%
  mutate(Methyl_factor=factor(`Methylation summary measure type`, 
                              levels=all_methyl_levels))%>%
  
  mutate(exposed=paste0(smkPreg_binary, ' prenatal smoke exposed & ', PostnatalMaternalSmokingAny))%>% 
  filter(Methyl_factor%in% main_methyl[3:4])%>%
  ggplot(aes(x=exposed, y=`Methylation summary measure`, color=exposed))+
    geom_violin()+geom_boxplot(width=0.2, position = position_dodge(0.9))+
  stat_summary(fun.data = give.n, geom = "text", fun.y = max,
                  position = position_dodge(width = 0.75))+
  stat_compare_means(comparisons=my_comparisons)+
    facet_grid(Methyl_factor~childteen, scales='free',labeller = labeller(Methyl_factor=all_methyl_labels))+
    scale_x_discrete(labels=function(x) str_wrap(x, width = 20))+
    theme(legend.position = 'none')+
  theme(axis.text.x = element_text(angle=30, hjust=1))+
  guides(color=guide_legend(nrow=4,byrow=TRUE))

ggexport(dna_prepost, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=10, height=10)
sup_fig=sup_fig+1
```

```{r supplemental figure pre post smoke exposure print, eval=as_supp, fig.width=14, fig.height=10}
dna_prepost
```

# Supplemental Figure `r sup_fig`: Heatmap of *P* values from Pearson correlations between known covariates and surrogate variables from surrogate variable models

```{r surrogate variables heatmap}
#bring the heatmap function
covar=c('Sample_Plate', 'Slide', 'Array', 'cm1bsex', 'smkPreg_binary', 'childteen', 'ancestry')
modeldata$Sample_Plate=set_label(modeldata$Sample_Plate, 'Sample Plate')
modeldata$cm1bsex=set_label(modeldata$cm1bsex, 'Child gender')
modeldata$PostnatalMaternalSmokingAny=set_label(modeldata$PostnatalMaternalSmokingAny, 'Postnatal maternal smoking (age 1 & 5)')
modeldata$SmkAtVisitPastmonth=set_label(modeldata$SmkAtVisitPastmonth, 'Postnatal maternal smoking (month prior to DNA methylation sample')

bring.the.heatmap.sv <- function(pd,covar,folder){
  library('gtools')
  nsv <- length(grep('sv',names(pd)))
  
  #generate correlation matrix between SVs and variables
  correlations <- matrix(0,nrow=length(covar),ncol=nsv)
  colnames(correlations) <- paste('sv',1:nsv,sep='')
  rownames(correlations) <- covar
  
  #take svs
  svs <- pd[,colnames(correlations)]
  
  #correlations between svs and variables:
  # cor test if continuous
  # anova is multi category
  # t-test if two category
  for(col in covar){
    var <- pd[,col]
    #check to see if it is numeric and isn't really a two category thing
    if(class(var)=='numeric' & dim(table(var))>2){
      correlations[col,] <- apply(svs,2,FUN = function(x,y) cor.test(x,y)$p.value,y=var)
    }else{
      if(length(levels(factor(var)))>2){
        correlations[col,] <- apply(svs,2,FUN = function(x,y) anova(lm(x~y))$'Pr(>F)'[1],y=var)
      }else{
        correlations[col,] <- apply(svs,2,FUN = function(x,y) t.test(x~y)$p.value,y=var)
      }
    }
  }
  color.gradient <- colorRampPalette(c('firebrick','ivory3','royalblue'),bias=3)
  correlations=correlations%>%as.data.frame()%>%tibble::rownames_to_column('var1')%>%
    pivot_longer(-var1, names_to='var2')%>%
    mutate(var2=factor(var2, levels=unique(mixedsort(var2))),
           var1=factor(var1, levels=pd[, covar]%>%colnames(), labels=get_label(pd[, covar])))
  correlations%>%
    ggplot(aes(x=var1, y=var2, fill=value))+geom_tile()+
    scale_fill_gradientn('P value', colours = color.gradient(20), values = seq(0, 1, 0.5))+
    scale_x_discrete('', labels=str_wrap(get_label(pd[, covar]), 10))+
    scale_y_discrete('')
}

pdf(file=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=11, height=13)
bring.the.heatmap.sv(modeldata, covar=c('smkPreg_binary', 'Sample_Plate', 'Slide', 'Array', 'ChildAgeComposite', 'ancestry', 'cm1bsex', 'Leukocytes_saliva', 'Epithelial.cells_saliva', 'cm1inpov', 'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'PostnatalMaternalSmokingAny', 'SmkAtVisitPastmonth'))
dev.off()
sup_fig=sup_fig+1
```

```{r surrogate variable heatmap plot, eval=as_supp, fig.width=12, fig.height=8}
bring.the.heatmap.sv(modeldata, covar=c('smkPreg_binary', 'Sample_Plate', 'Slide', 'Array', 'ChildAgeComposite', 'ancestry', 'cm1bsex', 'Leukocytes_saliva', 'Epithelial.cells_saliva', 'cm1inpov', 'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'PostnatalMaternalSmokingAny', 'SmkAtVisitPastmonth'))
```

# Supplemental Table `r sup_tab`: Coefficients for multivariable models of prenatal maternal smoke exposure regressed on selected DNA methylation summary variables (outcomes) among children in the Fragile Families and Child Wellbeing study 

```{r}

coefficient_table=rbind(global_models%>%
                          mutate(ancestry='Multiethnic sample'), 
                        local_models)%>%unnest(tidied)%>%
  filter(term=='smkPreg_binaryYes')%>%
  mutate(N=map_dbl(model, nobs))%>%
  dplyr::select(outcome, childteen, ancestry, predictors,
                N, estimate, conf.low, conf.high, p.value)%>%
  mutate(across(estimate:conf.high, ~signif(.x, 2)))%>%
  mutate(p.value=signif(p.value, 5))%>%
  arrange(outcome, childteen, ancestry, predictors)

  
write.csv(coefficient_table, file.path(outdir, paste0('SupplementalTable', sup_tab, '.csv')))
sup_tab=sup_tab+1

```


# Supplemental Figure `r sup_fig`: Associations between prenatal maternal smoking and polymethylation scores calculated using alternative coefficient sets among `n n_total` children in the Fragile Families and Child Wellbeing study. Base model: Child sex, maternal income to poverty ratio, immune cell proportion, plate, 1st two principal components from PCA on child genetic data, child age. Prenatal exposures model: base model + yes/no any prenatal maternal alcohol use, yes/no any prenatal other drug use. Secondhand smoking models: prenatal exposures model + yes/no any postnatal maternal/primary care giver smoking at age 1 or 5, pk/day in month prior to primary caregiver/maternal interview at age 9 or 15. Surrogate variable: including all surrogate variables


```{r supplemental figure compare across pms coefficients} 
 compare_pms=all_models%>%
    mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
    unnest(tidied)%>%
    filter(term=='smkPreg_binaryYes')%>%
    filter(str_detect(outcome, '28 site|19 site|568 site|204 site|cell-type'))%>%
  filter(ancestry=='Unstratified')%>%
 ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, rev(levels(age))), y=gsub('- PMC.*', '', outcome)))+
    geom_point(position = position_dodge(width = 1))+
    geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
    facet_wrap(predictors~., scales='free_x', labeller=labeller(predictors=label_wrap_gen(20)))+
    geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
    theme_classic()+
  scale_color_discrete(name='')+scale_x_continuous("Estimate (95% CI)")+scale_y_discrete('')

ggexport(compare_pms, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=18)
sup_fig=sup_fig+1

```

```{r compare across pms print, eval=as_supp, fig.width=10, fig.height=12}
compare_pms
```

# Supplemental Figure `r sup_fig` - Associations between prenatal maternal smoking and other *a priori* CpG sites selected from previous meta-analyses among `r n_total` children in the Fragile Families and Child Wellbeing study 

```{r supplemental figure compare to top cpgs}
topCPGs<-read.csv(paste0(datadir, "/TopCpGsWiklundEtAl2019.csv"))
topCPGs<-topCPGs %>% mutate(estimate=as.numeric(gsub("−\\s", "-", gsub(" .*", "", β..SE.))), std.error=as.numeric(sub("\\).*", "", sub(".*\\(", "", β..SE.))), p.value=P.value, source="Wiklund et al. 2019", conf.low=(estimate-1.96*std.error), conf.high=(estimate+1.96*std.error), set ='')%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)%>%mutate(estimate=estimate*100, conf.low=100*conf.low, conf.high=100*conf.high)

#read in Joubert meta top cpgs 
#read in cpgs from PMC4833289
cpgs<-read.csv(paste0(datadir, '/PMC4833289_SuppFile2CpGs.csv'), stringsAsFactors = F, header=F, skip=2)
#make headers
header1<-scan(paste0(datadir, '/PMC4833289_SuppFile2CpGs.csv'), nlines=1, what=character(), sep=',')
header1[7:26]<-c(rep("SSnewborn", 5), rep("SSnewbornCT", 5), rep("SSolder", 5), rep("anynewborn", 5))
header2<-paste(header1, scan(paste0(datadir, '/PMC4833289_SuppFile2CpGs.csv'), nlines=1, skip=1, what=character(), sep=','), sep='_')
names(cpgs)<-header2

#four meta-EWASes: SS_newborn (sustained smoking in newborns); SS_newbornCT (sustained smoking in newborns controlling for cell type) SS_older (sustained smoking in older children); any_newborn (any smoking in newborns)
#pivot_longer to get single column of coefs and a column for the analysis type
#and group into a list by analysis type
cpgs<-cpgs %>% pivot_longer(cols = c(matches("_Coef|_SE|_P|_FDR|_Direction")), names_to = c("set", ".value"), names_pattern="(.+)_(.+)")%>%rename_with(~gsub('_', '', .x))%>%mutate(Gene=`Mapped Gene`, estimate=Coef, conf.low=(estimate-1.96*SE), conf.high=(estimate+1.96*SE), source='Joubert et al. 2016', std.error=SE, p.value=P, Chr=CHR) %>%mutate(source=paste0(source, ': ', set))%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)%>%filter(set %in% c('SSolder', 'SSnewbornCT'))%>%mutate(estimate=estimate*100, conf.low=100*conf.low, conf.high=100*conf.high)

myCpGs=all_models%>%filter(str_detect(outcome, 'cg') & predictors=='base model')%>%unnest(tidied)%>%filter(term=='smkPreg_binaryYes')%>%
    filter(ancestry=='Unstratified')%>%mutate(CpG=gsub( '.*: ', '', outcome), Gene=gsub(':.*', '', outcome), source=paste0('Fragile Families:', age))

top_cpgs=myCpGs$CpG

cpg_all<-rbind(myCpGs%>%filter(CpG %in% top_cpgs), cpgs%>%filter(CpG %in% top_cpgs)%>%dplyr::select(-Chr), topCPGs%>%filter(CpG%in%top_cpgs)%>%mutate_at(c('p.value'), as.numeric))


#geom blank to give good axes limits
blank_data <- cpg_all %>%dplyr::select(Gene, CpG, estimate, std.error)%>%group_by(Gene)%>%mutate(min=min(estimate), max=max(as.numeric(std.error), na.rm=T))%>%dplyr::select(-estimate, -std.error)%>%unique()%>%mutate(estimate=min*-1+max, conf.high=estimate, conf.low=estimate)%>%dplyr::select(-max, -min)%>%arrange(Gene)%>%mutate(source='Wiklund et al. 2019')


blank_data2 <- cpg_all %>%dplyr::select(Gene, CpG, estimate, std.error)%>%group_by(Gene)%>%mutate(min=min(estimate))%>%dplyr::select(-estimate, -std.error)%>%unique()%>%mutate(estimate=min*-1, conf.high=estimate, conf.low=estimate)%>%dplyr::select(-min)%>%arrange(Gene)%>%mutate(source='Fragile Families: Age_9')

plot_aprioriCpG<-ggplot(cpg_all, aes(x=estimate, xmin=conf.low, xmax=conf.high, color=source, y=CpG))+geom_point(position=position_dodge(width=1))+geom_errorbarh(height=0.1, position=position_dodge(width=1))+scale_x_continuous()+geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+facet_wrap(.~Gene, scales='free', ncol=2)+theme_classic()+theme(legend.position = 'bottom')+guides(color=guide_legend(nrow=4,byrow=TRUE))+geom_blank(data = blank_data, aes(x = estimate))+scale_color_discrete('', labels=c('Fragile families saliva: age 15', 'Fragile families saliva: age 9', 'Joubert et al. 2016 newborn cord blood', 'Joubert et al. 2016 peripheral blood older children', 'Wicklund et al. peripheral blood >16 years'))+scale_x_continuous('Estimate (95% CI)')

ggexport(plot_aprioriCpG, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=9)
sup_fig=sup_fig+1
```

```{r compare to top cpgs print, eval=as_supp, fig.width=10}
plot_aprioriCpG
```

# Main figure `r main_fig`: Area under the curve and receiver operating curve for the polymethylation score, *AHRR*:cg05575921, and global methylation

```{r main figure 4 auc}

age_labels=c('Age 9', 'Age 15'); names(age_labels)=c('Age 9', 'Age 15')
methyl_labels_roc=paste0('Base model+', all_methyl_labels)
names(methyl_labels_roc)=paste0('Base model+', names(all_methyl_labels))

fig4a= global_roc%>%
    dplyr::select(predictors, covariates, methylation, childteen, test, modeltype)%>%
    unnest(test) %>% 
    filter(str_detect(methylation, 'SSnewbornCT|cg0557|global') | modeltype=='Base model')%>%
  mutate(modeltype=gsub(": coefficients.*)", " for prenatal smoke exposure (Joubert cell-type control coefficients)", modeltype))%>%
  mutate(modeltype=factor(modeltype, levels=c('AHRR: cg05575921', 'Base model+AHRR: cg05575921', 'Global methylation', 'Base model+Global methylation',  '6074 site polymethylation score for prenatal smoke exposure (Joubert cell-type control coefficients)', 'Base model+6074 site polymethylation score for prenatal smoke exposure (Joubert cell-type control coefficients)')))%>%
    ggplot(aes(x=spec, y=sens, color=modeltype))+geom_line()+
    facet_wrap(childteen~., nrow=1, labeller=labeller(childteen=age_labels))+
    theme_classic()+
    scale_color_manual('', values=c('red', 'darkred', 'green', 'darkgreen', 'steelblue1', 'royalblue3'), 
                       function(x) str_wrap(x, width = 30))+
    theme(legend.position = 'right')+guides(color=guide_legend(nrow=4, byrow=TRUE))+
    geom_abline(intercept=1, slope=-1, color='black')+
    xlab('Specificity')+ylab('Sensitivity')+
    theme(legend.position='bottom')+ coord_fixed()

fig4b=global_roc%>%
  dplyr::select(covariates, methylation, childteen, test)%>% 
  filter(str_detect(methylation, 'SSnewbornCT') & covariates!='')%>%unnest(test)%>%
  rbind(long_global_roc%>% 
          mutate(childteen='Age 9 & Age 15', methylation=gsub('_Age_9.*', '', methylation))%>%
          dplyr::select(covariates, methylation, childteen, test)%>% 
          filter(str_detect(methylation, 'SSnewbornCT') 
                 & covariates!='')%>%
          unnest(test))%>%
  ggplot(aes(x=spec, y=sens, color=childteen, linetype=childteen))+geom_line(size=2)+
  facet_wrap(~methylation, labeller = labeller(methylation=c('SSnewbornCT_center_scale'='Polymethylation score')))+
  theme_classic()+theme(legend.position = 'right')+guides(color=guide_legend(nrow=4,byrow=TRUE))+
  geom_abline(intercept=1, slope=-1, color='black')+
  scale_color_grey('', labels=c('Age 9 only', 'Age 9 & 15', 'Age 15 only'))+
  scale_linetype('', labels=c('Age 9 only', 'Age 9 & 15', 'Age 15 only'))+
  xlab('Specificity')+ylab('Sensitivity')+
  theme(legend.position='bottom')+ coord_fixed()

auc_fig=cowplot::plot_grid(fig4a, fig4b, labels=c('A', 'B'), nrow=1, rel_widths=c(0.63, 0.37), align='h', axis='tb')

ggexport(auc_fig, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=9)
ggexport(auc_fig, filename=file.path(outdir, paste0('Figure', main_fig, '.eps')), width=9)
main_fig=main_fig+1
```

```{r main figure 4 auc print, eval=stand_alone, fig.width=12}
auc_fig
```

# Supplemental Table `r sup_tab`: Area under the curve for models shown in Main Figure 4 for classifying prenatal maternal smoke exposure, with *P* value when comparing to base model only or to model using polymethylation score + base model. Base model: Child sex, maternal income to poverty ratio, immune cell proportion, plate, 1st two principal components from PCA on child genetic data, child age.  

```{r supplemental table auc}
#CI.delong is the confidence interval for each auc: if we sampled 100 times, 95 of the samples would have a auc between those two numbers
roc_table=global_roc %>% 
  dplyr::select(childteen, methylation, modeltype, roc, auc, ci_auc_delong)%>%
  mutate(ci.delong=map_chr(ci_auc_delong, ~paste(round(.x[1], 2), round(.x[3], 2), sep=', ')))

roc_base=roc_table %>% 
  filter(modeltype=='Base model+No methylation')%>%
  rename_with(~str_c(., '_basemodel'), .cols=c(roc, auc, ci.delong))%>%
  dplyr::select(childteen, contains('_basemodel'))

roc_ss=roc_table %>% 
  filter(modeltype=='Base model+Sustained smoking polymethylation score (newborns, cell-type controlled)')%>%
  rename_with(~str_c(., '_pms'), .cols=c(roc, auc, ci.delong))%>%
  dplyr::select(childteen, contains('_pms'))



#confidence_interval_difference_pms|base is the confidence interval for the DIFFERENCE between the two aucs - avoids incorrect error propagation, since the squares of errors add, not the errors themselves (see an explanation here: https://towardsdatascience.com/why-overlapping-confidence-intervals-mean-nothing-about-statistical-significance-48360559900a)
#delong test is for nested models. the base model is always nested within the base model+any of the methylation measures 
rct1=roc_table%>% 
  filter(methylation %in% c('globalmethylation', 'cg05575921', 'SSnewbornCT_center_scale', ''))%>%
  filter(str_detect(modeltype, 'Base model'))%>%
  left_join(roc_base)%>%
  mutate(roctest_base=map2(roc, roc_basemodel, ~roc.test(.x, .y, method='delong')))%>%
  mutate(auc=round(auc, 2))%>%
  mutate(`P value vs base model (DeLong)`=map_dbl(roctest_base, ~round(.x$p.value, 4)))%>%
  mutate(`Difference (AUC - AUC of base model) (95% CI - DeLong)`= map_chr(roctest_base,
                   ~paste0(round(.x$estimate[1]-.x$estimate[2], 2), ' (95% CI: (', 
                           round(.x$conf.int[1], 3), ', ', 
                           round(.x$conf.int[2], 3), '))')))%>%
  mutate(`Percentage Difference (AUC - AUC of base model) (95% CI - DeLong)`= map_chr(roctest_base,
                   ~paste0(round(.x$estimate[1]-.x$estimate[2], 2)*100, ' (95% CI: (', 
                           round(.x$conf.int[1], 3)*100, ', ', 
                           round(.x$conf.int[2], 3)*100, '))')))%>%
  mutate(auc_basemodel=round(auc_basemodel, 2))%>%
     dplyr::select(childteen, modeltype, auc, auc_basemodel,
                `Percentage Difference (AUC - AUC of base model) (95% CI - DeLong)`,
                `Difference (AUC - AUC of base model) (95% CI - DeLong)`, `P value vs base model (DeLong)`)
  
rct1_export=rct1%>%
  dplyr::select(childteen, modeltype, auc, auc_basemodel,
                `Difference (AUC - AUC of base model) (95% CI - DeLong)`, `P value vs base model (DeLong)`)

write.csv(rct1_export, file=file.path(outdir, paste0('SupplementalTable', sup_tab, '.csv')), row.names=F)

sup_tab=sup_tab+1  
```

```{r supplemental table auc print, eval=as_supp}
rct1_export%>%dplyr::filter(!str_detect(modeltype,'No methylation'))%>%kbl(col.names=c('Age', 'Model', 'AUC', 'Base model AUC', 'Difference (AUC - AUC of base model) (95% CI - DeLong)', 'P value for difference between AUC and base model AUC (DeLong)'))
```

```{r}

all_rocs=roc_table %>% 
  filter(str_detect(modeltype, '19 site|204 site|568 site|cell-type|AHRR|net|global|PedBE'))%>%
  arrange(childteen, methylation)%>%
  dplyr::select(childteen, modeltype, auc, ci.delong)

write.csv(all_rocs, file=file.path(outdir, paste0('SupplementalTable', sup_tab, '.csv')))
sup_tab=sup_tab+1
```


# Main Figure `r main_fig` forest plot comparison of AUCs for the different polymethylation scores calculated in Fagile Families saliva DNA methylation samples versus in cord and peripheral blood samples of other cohorts 


```{r auc comparison across studies main figure 5}
#change ! to no ! to use model with additional covariates 
my_aucs=global_roc %>% 
  filter(str_detect(modeltype, 'Base model') & methylation%in%c('SSnewbornCT_center_scale', 'reese_center_scale','richmond19_center_scale', 'richmond568_center_scale',  'elastic_net_score'))%>%
  mutate(l.ci=map_dbl(ci_auc_delong, ~.x[[1]]), u.ci=map_dbl(ci_auc_delong, ~.x[[3]]),
         cohort=paste0(childteen, ' - Fragile Families', ' (saliva)'), 
         PMID=NA, exposure='self-report', 
         method=case_when(methylation=='SSnewbornCT_center_scale'~
                            '6074 site polymethylation score (Joubert newborn cord blood coefficients (cell-type corrected))', 
                          methylation=='reese_center_scale'~
                            '28 site polymethylation score (Reese newborn cord blood LASSO coefficients))', 
                          methylation=='elastic_net_score'~
                            '204 site polymethylation score (Rauschert adolescent peripheral blood elastic net coefficients)', 
                          methylation == 'richmond19_center_scale'~
                          '19 site polymethylation score (Joubert older children peripheral blood coefficients (Richmond))', 
                          methylation == 'richmond568_center_scale'~
                            '568 site polymethylation score (Joubert newborn cord blood coefficients (Richmond))'),
         age=as.numeric(word(cohort, 2)))%>%
  dplyr::select(cohort, auc, l.ci, u.ci, PMID, exposure, method)


other_aucs=data.frame(auc=c(0.87, 0.88, 0.83, 0.85, 0.85, 0.72, 0.73,
                            0.85, 0.74, 0.54, 0.79, 0.8, 0.71, 0.73, 
                            0.8, 0.68, 0.48, 0.75, 0.73, 0.72, 0.73,
                            0.69, 0.72, 0.865,  0.90), 
                      l.ci=c(rep(NA, 21), 0.67, 0.69, NA, 0.83), 
                      u.ci=c(rep(NA, 21), 0.73, 0.76, NA, 0.97), 
                      cohort=c(rep('Age 17 - Raine Study (peripheral blood)', 7),
                               rep('Age 16 - NFBC1986 (peripheral blood)', 7), 
                               rep('Age 31 - NFBC1966 (peripheral blood)', 7), 
                               'Age 30 - ALSPAC women (peripheral blood)',
                               'Age 30 - ALSPAC women (peripheral blood)',
                               'Age 3-5 - SEED children (peripheral blood)', 
                               #'Age 0 - MoBa newborns (cord blood)',
                               'Age 0 - MoBa newborns (cord blood)'),
                      exposure=c(rep('self-report', 21), 'self-report', 'self-report', 'self-report', 'combined'),
                      method=c(rep(c('204 site polymethylation score (Rauschert adolescent peripheral blood elastic net coefficients)', 'gradient boosting machine', 
                                     'random forest', 'support vector machine', 
                               '28 site polymethylation score (Reese newborn cord blood LASSO coefficients))',
                               '568 site polymethylation score (Joubert newborn cord blood coefficients (Richmond))', 
                               '19 site polymethylation score (Joubert older children peripheral blood coefficients (Richmond))'), 3), 
                         '568 site polymethylation score (Joubert newborn cord blood coefficients (Richmond))', 
                               '19 site polymethylation score (Joubert older children peripheral blood coefficients (Richmond))',
                        'Support vector machine', 
                        #'28 site polymethylation score (Reese newborn cord blood LASSO coefficients))',
                        '28 site polymethylation score (Reese newborn cord blood LASSO coefficients))'),
                      PMID=c(rep(32930613, 21), 
                             29860346, 
                             26610292,
                             26610292, 
                             #27323799,
                             27323799))%>% 
                      mutate(age=as.numeric(word(cohort, 2, sep=' |-')))

all_aucs=rbind(my_aucs, other_aucs)%>%
         mutate(cohort=factor(cohort, levels=c('Age 0 - MoBa newborns (cord blood)', 
                                               'Age 3-5 - SEED children (peripheral blood)', 
                                               "Age 9 - Fragile Families (saliva)" , 
                                               "Age 15 - Fragile Families (saliva)", 
                                               'Age 16 - NFBC1986 (peripheral blood)',
                                               'Age 17 - Raine Study (peripheral blood)', 
                                               'Age 30 - ALSPAC women (peripheral blood)', 
                                               'Age 31 - NFBC1966 (peripheral blood)')))%>%
  mutate(cohort2=factor(cohort, levels=rev(levels(cohort))))

auc_compare=ggplot(all_aucs%>%filter(str_detect(method, 'score')), 
                   aes(y=cohort2, x=auc, xmin=l.ci, xmax=u.ci, color=cohort2))+
  geom_point(position=position_dodge(width=1))+
  geom_errorbarh(height=0.1, position=position_dodge(width=1))+
  scale_x_continuous()+
  #scale_y_discrete(limits = rev(levels(all_aucs$cohort)))+
  geom_vline(xintercept=0.5, color='red', linetype="dashed", alpha=0.5)+
  theme(legend.position='none')+ggforce::facet_col(vars(method), scales='free_y', space='free', strip.position='top')
ggexport(auc_compare, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=9)
ggexport(auc_compare, filename=file.path(outdir, paste0('Figure', main_fig, '.eps')), width=12)

```








