---
title: "Polymethylation scores results"
author: "F blostein"
date: "9/30/2021"
output:
  html_document:
    code_folding: show
    highlight: tango
    number_sections: no
    theme: sandstone
  pdf_document: default
  word_document: default
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message=FALSE)

library(ggplot2)
library(kableExtra)
library(ggpubr)
library(tidyr)
library(stringr)
library(dplyr)
library(compareGroups)
library(kableExtra)
library(gtsummary)
library(gtools)
library(haven)
library(sjlabelled)
library(matchmaker)
library(EpiDISH)
library("PerformanceAnalytics")
library(DescTools)
library(purrr)
library(broom)
library(broom.mixed)
library(ggplot2)
library(multcomp)
library(ggpubr)
library(cowplot)
library(geepack)
library(lme4)
library(pROC)
library(lmerTest)
library(stringr)
library(sva)
library(gtsummary)
library(gt)
library(here)
library(DiagrammeRsvg)
library(xml2)
library(ggrepel)

include_ewas=F
stand_alone=F
```

```{r temp read n data}
basedir=here::here()
outdir=file.path(gsub('Code', 'Output', basedir), 'Figures')
datadir=gsub('Code', 'Data', basedir)
load(paste0(datadir, '/CreatedData/allPhenoData.Rdata'))
load(paste0(datadir, '/CreatedData/allmethyl.Rdata'))
load(paste0(datadir, '/CreatedData/completeCaseSVA.Rdata'))
load(paste0(datadir, '/CreatedData/methylation_summarymodels.Rdata'))
```

```{r plotting variables}
theme_set(theme_bw())
```

```{r}
modeldata=modeldata %>% mutate(childteen=factor(childteen, levels=c('Age 9', 'Age 15')))
set_label(modeldata$childteen)='Visit'

n_total=modeldata %>% dplyr::select(id)%>%unique()%>%nrow()
m_obs=modeldata%>%nrow()
n_twovisit=modeldata %>% group_by(id)%>%summarise(n=n())%>%filter(n==2)%>%nrow()

all_methyl=c("globalmethylation", "pediatric", "grim", 'DNAmPACKYRS', "Epithelial.cells_saliva", "Leukocytes_saliva", "SSnewborn_center", "SSnewbornCT_center", "SSolder_center", "cg05575921")
main_methyl=c("globalmethylation", "pediatric", "SSnewbornCT_center", "cg05575921")


all_methyl_labels=modeldata%>%dplyr::select(any_of(all_methyl))%>%get_label()
all_methyl_labels[str_detect(all_methyl_labels, 'Polymethylation', )]=paste0(rep('Score:', 3), c('Sustained smoking \n newborns', 'Sustained smoking \n newborns (w/ CT control)', 'Sustained smoking \n older children'))
names(all_methyl_labels)=all_methyl

all_methyl_levels=c("SSnewbornCT_center", "SSnewborn_center",  "SSolder_center",
                                       "pediatric" , "grim", 'DNAmPACKYRS',  "cg05575921", 
                                      "globalmethylation","Epithelial.cells_saliva", "Leukocytes_saliva")

main_methyl_labels=modeldata%>%dplyr::select(any_of(main_methyl))%>%get_label()

my_round=function(x, d){
  y=c()
  d_orig=d
  for(i in 1:length(x)){
    d=d_orig
    while(round(x[i], d)==0){
    d=d+1
    }
    y[i]=round(x[i], d)
  }
  return(y)
}
```

```{r figure counter}
main_fig=1
sup_fig=1
main_tab=1
sup_tab=1
```

# Figure `r main_fig` - Selection of samples from the Fragile Families and Child Wellbeing (FFCW) study into this analysis. N represents the number of individuals at each step in the selection procedure, M represents the number of samples, as individuals with repeated measures can have more than one sample.

```{r, eval=T}
#function
N_M_count=function(filter_statement, statement, df){
  paste0('m=', df %>% filter(eval(parse(text = filter_statement)))%>%nrow(), 
         ' samples & ', 
         'n=', df %>% filter(eval(parse(text = filter_statement))) %>% dplyr::select(idnum) %>%unique()%>%nrow(), 
         ' individuals with', statement)
}
#Make Labels
n_fullFF<-paste0("Fragile Families cohort \n", "n=", FF_labeled %>% nrow())
n_methylQC<-paste0("Cohort with 450K methylation data \n", "n=", methylCohort$idnum%>%unique()%>%length(), " individuals with m=", methylCohort$MethID%>%length(), " samples")

n_QCloss<-paste0("Methylation data QC: ", 
                 N_M_count("probe_fail_10==T & childteen!='M'", " with >10% of probes failing dropped, \n ", pdqc_all),
                 N_M_count("sex_new!=predicted_sex & probe_fail_10==F & childteen!='M'", 
                           "  with discordant sex samples dropped, \n ", pdqc_all),
                 N_M_count("predicted_sex_outlier==T & sex_new==predicted_sex & probe_fail_10==F & childteen!='M'", 
                           ' with abnormal sex chromosome intensity dropped & \n ', pdqc_all),
                 length(reps_remove), " duplicate samples dropped")

n_methylFF<-paste0("Cohort with quality controlled 450K data \n","n=", myFF %>% dplyr::select(id) %>% unique() %>% nrow(), " individuals with m=", myFF%>%nrow(), " samples")


n_missingmaternalprenataldata=paste0('Missing maternal data: \n', 
                                     N_M_count("is.na(smkPreg_binary)", " with missing prenatal maternal smoking \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & is.na(m1g2_YesNoPreg)", " with missing prenatal alcohol \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & is.na(m1g3_YesNoPreg)", 
                                               " with missing prenatal other drug use \n ", myFF),
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & is.na(cm1inpov)", 
                                               " with missing maternal income to poverty ratio data \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) &
                                               PostnatalMaternalSmokingAny=='Missing'", " with missing postnatal maternal smoking data \n ", myFF), 
                                     N_M_count("!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) &
                                               PostnatalMaternalSmokingAny!='Missing' & SmkAtVisitPastmonth=='Missing'", 
                                               " with missing past month maternal/primary care giver smoking packs/day", myFF))

maternaldata=myFF %>% filter(!is.na(smkPreg_binary) & !is.na(m1g2_YesNoPreg) & !is.na(m1g3_YesNoPreg) & !is.na(cm1inpov) & PostnatalMaternalSmokingAny!='Missing' & SmkAtVisitPastmonth!='Missing')  

n_withmaternaldata=paste0('m=', maternaldata %>% nrow(), ' samples & n=', maternaldata %>% dplyr::select(idnum)%>%unique()%>%nrow(), ' individuals with maternal covariate data')
  
n_missingchilddata=paste0('Missing child data: \n', 
                          N_M_count("is.na(ancestry) | ancestry=='Missing PC data'", 
                                    " with missing ancestry/genetic data  \n ", maternaldata), 
                          N_M_count("!is.na(ancestry) & ancestry!='Missing PC data' & is.na(cm1bsex)", 
                                    " with missing child sex data \n ", maternaldata), 
                          N_M_count("is.na(ancestry) | ancestry=='Missing PC data' & !is.na(cm1bsex) & is.na(ChildAgeComposite)", 
                                    " with missing child age at sample \n ", maternaldata))

childdata= maternaldata %>% filter(ancestry!='Missing PC data' & !is.na(cm1bsex) & !is.na(ChildAgeComposite))
n_withchilddata=paste0('m=', childdata %>% nrow(), ' samples & n=', childdata %>% dplyr::select(idnum)%>%unique()%>%nrow(), ' individuals with child covariate data')

n_Child<-paste0('n=', childdata %>%filter(childteen=='C')  %>% nrow(), " individuals with a 9 year visit")
n_ChildSmkexcluded<-paste0("Exclude children who smoke: \n",'n=', childdata%>%filter(childteen=='C' & k5f1l=='2 no')%>% nrow(), 
                           " individuals with a 9 year visit")
n_ChildSmkexclusions<-paste0('n=', childdata %>% filter((k5f1l=="1 yes" | is.na(k5f1l)) & childteen=='C') %>% nrow(), 
                             " individuals smoking \n or missing child smoking data \n at 9 year visit")
n_Teen<-paste0('n=', childdata %>%filter(childteen=='T')   %>% nrow(), " individuals with a 15 year visit")
n_TeenSmkexcluded<-paste0("Exclude children who smoke: \n",'n=', childdata %>%filter(childteen=='T' & k6d40=="2 No") %>% nrow(), 
                          " individuals with a 15 year visit")
n_TeenSmkexclusions<-paste0('n=', childdata %>% filter((k6d40=="1 Yes" | is.na(k6d40) | k5f1l=='1 yes') & childteen=='T') %>% nrow,
                            " individuals smoking \n or missing focal child smoking data \n at 15 year visit") 


if(childdata %>% filter(((k5f1l=="2 no" | (is.na(k5f1l) & k6d40=='2 No')) & childteen=='C') | (childteen=='T' & k6d40=="2 No" & (k5f1l=='2 no' |  is.na(k5f1l))))%>%nrow()!=nrow(modeldata)){stop('filter N and completecase data set N do not match')}


```

```{r, fig.height=12}

graph1<-DiagrammeR::grViz("
   digraph graph2 {
   compound=true;
   graph [layout = dot]
   
   node [shape = rectangle, width=4]
   a [label = '@@1']
   b [label = '@@2']
   c [label = '@@3']
   d [label = '@@4']
   e [label = '@@5']
   f [label = '@@6']
   g [label = '@@7']
   h [label = '@@11']
   i [label = '@@12']
   

   #fake nodes for connecting to edges
   node [shape=point, width=0.01, height=0.01]
   a1
   b1
   c1
   d1
   f1 [group=g1]
   g1
   h1
  
  #exclusion nodes
  node [shape = rectangle, width=4]
  aa [label = '@@8']
  bb [label = '@@9']
  cc [label = '@@10']
  hh [label = '@@13']
  ii [label = '@@14']
    
  #draw connections
#  a->m
#  m->a1 [arrowhead=none]
  a->a1 [arrowhead=none]
  a1->b
  b->b1 [arrowhead=none]
  b1->c
  c->c1 [arrowhead=none]
  c1->d
  d->d1 [arrowhead=none]
  d1->e
  e->f; e->g
  f->g1 [arrowhead=none]; g->h1 [arrowhead=none]
  g1->h; h1->i 
  
  
{rank=same; b1->aa [minlen=2]}   
{rank=same; c1->bb [minlen=2]} 
{rank=same; d1->cc [minlen=2]}
{rank=same; g1->hh [minlen=2]}
{rank=same; h1->ii [minlen=2]}



#invisible constraints to force ordering
#ff->i [style=invis]
   }   
   
   [1]: n_fullFF
   [2]: n_methylQC
   [3]: n_methylFF
   [4]: n_withmaternaldata
   [5]: n_withchilddata
   [6]: n_Child
   [7]: n_Teen
   [8]: n_QCloss
   [9]: n_missingmaternalprenataldata
   [10]: n_missingchilddata
   [11]: n_ChildSmkexcluded
   [12]: n_TeenSmkexcluded
   [13]: n_ChildSmkexclusions
   [14]: n_TeenSmkexclusions
")

graph1 %>%
  export_svg() %>%
  read_xml() %>%
  write_xml(file.path(outdir, paste0("Figure", main_fig, ".svg")))

main_fig=main_fig+1            
```
```{r, eval=stand_alone}
graph1
```

# Supplemental Table `r sup_tab` - Descriptive statistics of excluded vs included samples with methylation data available from the Fragile and Families and Child Wellbeing study

```{r}
myFF$MethylData2<-ifelse(myFF$MethID %in% modeldata$MethID, 'Final analysis group', 'Not in final analysis group')
set_label(myFF$MethylData2)<-'Exclusions based on missing data'
myFF=left_join(myFF, methyldata%>%mutate(idnum=as.character(idnum)))
myFF=myFF%>%mutate(across(c(PostnatalMaternalSmokingAny, SmkAtVisitPastmonth), ~ifelse(.x=='Missing', 
                                                       NA, .x)))%>%
  mutate(ancestry=ifelse(ancestry=='Missing PC data', NA, ancestry))%>%copy_labels(myFF)

exclusion_tbl=myFF %>% mutate(childteen=case_when(childteen=='C'~'Age 9', childteen=='T'~'Age 15'))%>%
  dplyr::select(smkPreg_binary, m1g2_YesNoPreg, m1g3_YesNoPreg, cm1inpov, PostnatalMaternalSmokingAny,
         SmkAtVisitPastmonth, ancestry, cm1bsex, childteen, globalmethylation, pediatric, grim,
         Epithelial.cells_saliva, Leukocytes_saliva, SSnewbornCT_center, SSolder_center,
         cg05575921, MethylData2)%>%
  tbl_summary(., by=MethylData2, statistic=list(all_continuous()~"{mean} ({sd})"))%>%
  modify_header(stat_by = "**{level}**, M samples = {n}")%>%
  add_n(col_label='M samples')#%>%
  #add_p(test=list(all_continuous()~"t.test"))

sup_tab=sup_tab+1
```

```{r, eval=stand_alone}
exclusion_tbl
```

```{r wide data}
modeldata_wide = modeldata%>%
  mutate(childteen=gsub(' ', '_', childteen))%>%
  dplyr::select(any_of(c('idnum', 'ancestry', 'childteen', 'cm1bsex', 'ancestry', 
                         'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'cm1inpov',
                         'smkPreg_binary',  all_methyl)))%>%
  pivot_wider(names_from = childteen, values_from=any_of(all_methyl))

model_twovisit=modeldata_wide[complete.cases(modeldata_wide), ]

modeldata_people=modeldata%>%
  dplyr::select(any_of(c('idnum', 'ancestry', 'cm1bsex', 'ancestry', 
                         'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'cm1inpov',
                         'smkPreg_binary', 'PostnatalMaternalSmokingAny')))%>%unique()
```


# Supplemental Figure `r sup_fig` - Pearson correlation of methylation summary measures from an anlysis of M=`r m_obs` samples and N=`r n_total`children in the Fragile Families and Child Wellbeing cohort

```{r}
library(Hmisc)
library(corrplot)
res2<-modeldata%>%dplyr::select(any_of(all_methyl))%>%cor()
#labels
colnames(res2)=all_methyl_labels
rownames(res2)=all_methyl_labels

pdf(file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
corrplot.mixed(res2, upper="ellipse", lower='number', order='hclust', tl.pos ="lt", tl.cex=0.8)
dev.off()

sup_fig=sup_fig+1
```

```{r, eval=stand_alone, fig.width=12, fig.height=12}
corrplot.mixed(res2, upper="ellipse", lower='number', order='hclust', tl.pos ="lt", tl.cex=0.8)

```

# Supplemental Figure `r sup_fig` - Correlation between age 9 and age 15 visits for selected DNA methylation summary measures among `r n_twovisit` children with age 9 and age 15 visit data in the Fragile Families and Child Wellbeing study

```{r}
age_long_methyl=modeldata%>%
  dplyr::select(any_of(c('idnum', 'childteen', 'ancestry', 'smkPreg_binary', all_methyl)))%>%
  pivot_longer(cols=any_of(all_methyl),
               names_to='Methylation summary measure type',
               values_to='Methylation summary measure')%>%
  mutate(Methyl_factor=factor(`Methylation summary measure type`, 
                              levels=all_methyl_levels))

age_wide_methyl=age_long_methyl%>%
  pivot_wider(names_from=childteen, 
              values_from=`Methylation summary measure`)

#checks
if(age_wide_methyl%>%filter(Methyl_factor=='globalmethylation')%>%filter(!is.na(`Age 9`) & !is.na(`Age 15`))%>%nrow()!=n_twovisit){stop('check sample size two visit vs age_wide methyl')}


visit_cor=age_wide_methyl%>%
  ggplot(aes(x=`Age 9`, y=`Age 15`))+geom_point()+stat_cor()+geom_smooth(method=lm)+theme_classic()+
  scale_x_continuous('Age 9 score')+scale_y_continuous('Age 15 score')+
  facet_wrap(~Methyl_factor, scales='free', labeller = labeller(Methyl_factor=all_methyl_labels), ncol=3)

ggexport(visit_cor, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1
```

```{r, eval=stand_alone, fig.width=12, fig.height=12}
visit_cor
```

# Supplemental Figure `r sup_fig`: Distributions of selected methylation summary measures at age 9 and age 15 among `r n_total` children in the Fragile Families and Child Wellbeing study

```{r}
dist_methyl=age_long_methyl%>%
  ggplot(aes(x=childteen, y=`Methylation summary measure`, color=childteen))+
  geom_violin()+geom_boxplot(width=0.2)+stat_compare_means(label="..p.signif..", label.y.npc=0.87)+
  scale_color_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  scale_x_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  geom_point(alpha=0)+geom_line(alpha=0, aes(group=idnum))+geom_smooth(aes(group=1), method='lm', color='black')+
  facet_wrap(~Methyl_factor, scales='free', labeller = labeller(Methyl_factor=all_methyl_labels), ncol=3)+
  theme(legend.position = 'bottom')


ggexport(dist_methyl, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1

```

```{r, eval=stand_alone, fig.height=11}
dist_methyl
```

# Figure `r main_fig`: Differences in selected DNA methylation summary measures by prenatal maternal smoke exposure among `r n_total` US children in the Fragile Families and Child Wellbeing study

```{r change in scores time, fig.width=12}

#add line for mean ? 
outcome_labels=c('Global methylation', 'Pediatric clock', 'AHRR: cg05575921', 'Polymethylation score \n (newborn, mean centered)', 'Polymethylation score  \n (older children, mean centered')
names(outcome_labels)=c('globalmethylation', 'pediatric', 'cg05575921', 'SSnewbornCT_center','SSolder_center')

fig2=age_long_methyl%>%
    filter(Methyl_factor%in% main_methyl)%>%
    ggplot(aes(x=childteen, y=`Methylation summary measure`, color=smkPreg_binary))+
    geom_violin()+geom_boxplot(width=0.2, position = position_dodge(0.9))+
  stat_compare_means(label='..p.signif..', label.y.npc = 0.88)+
    facet_wrap(~Methyl_factor, scales='free',labeller = labeller(Methyl_factor=all_methyl_labels))+
    scale_x_discrete('Visit', labels=c('Age 9', 'Age 15'))+
    scale_color_grey('', labels=c('Not prenatal maternal smoke exposed', 'Prenatal smoke exposed'))+
    theme(legend.position = 'bottom')




ggexport(fig2, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')))
main_fig=main_fig+1

```

```{r, eval=stand_alone}
fig2
```

# Table `r main_tab`: Bivariate associations between prenatal maternal smoking, selected DNA methylation summary measures and important covariates among a diverse sample of `r n_total` children in the Fragile Families and Child Wellbeing study

```{r bivariate table}
smoking_tbl=modeldata %>% 
  dplyr::select(smkPreg_binary, m1g2_YesNoPreg, m1g3_YesNoPreg, cm1inpov, PostnatalMaternalSmokingAny,
         SmkAtVisitPastmonth, ancestry, cm1bsex, globalmethylation, pediatric, grim, DNAmPACKYRS, Epithelial.cells_saliva,
         Leukocytes_saliva, SSnewbornCT_center, SSolder_center, cg05575921, childteen)%>%
  tbl_strata(strata=childteen, .tbl_fun=~.x %>%
               tbl_summary(by=smkPreg_binary, statistic=
                             list(all_continuous()~"{mean} ({sd})"))%>%
               add_n()%>%add_p(test=list(all_continuous()~"t.test")))

smoking_tbl=smoking_tbl%>%as_gt()%>%
  tab_row_group(label='Maternal', rows = 1:11)%>%
  tab_row_group(label='Child', rows=12:25)



main_tab=main_tab+1
```

```{r, eval=stand_alone}
smoking_tbl
```

# Figure `r main_fig` - Polymethylation scores are associated with prenatal maternal smoke exposure consistently across age and ancestry group in a sample of `r n_total` US children

```{r}
my_outcomes=c('Global methylation', 'Pediatric clock', 'AHRR: cg05575921', 'Sustained smoking polymethylation score (newborns, cell-type controlled)')

outcome_labels=c('Global methylation', 'Pediatric clock', 'AHRR: cg05575921',  'Polymethylation score \n (newborn, mean centered)')

all_models=rbind(local_models, global_models%>%mutate(ancestry='Unstratified'))

data_text=all_models%>%
  mutate(n=map_dbl(data, nrow), n=paste0('n=', n), x=-0.25)%>%
  unnest(tidied)%>%
  filter(outcome=='Sustained smoking polymethylation score (newborns, cell-type controlled)' & term=='smkPreg_binaryYes'& predictors=='secondhand smoke exposure model')%>%mutate(outcome_f=outcome)
  
names(outcome_labels)=my_outcomes

forest_plot=all_models%>%
  mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
  unnest(tidied)%>%
  filter(term=='smkPreg_binaryYes' & predictors=='secondhand smoke exposure model')%>%
  filter(outcome%in% my_outcomes)%>%
    mutate(outcome_f=factor(outcome, levels=rev(my_outcomes)))%>%
  ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, rev(levels(age))), y=ancestry))+
  geom_point(position = position_dodge(width = 1))+
  geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
  facet_grid(.~outcome_f, scales='free_x', labeller = labeller(outcome_f=outcome_labels))+
  geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
  geom_text(data=data_text, aes(x=x,label=n), position=position_dodge(width = 1), hjust=0.05)+
  theme_classic()+
  theme(text = element_text(size=16))+theme(legend.position = 'bottom')+
  scale_x_continuous('Coefficient for prenatal maternal smoking + 95% CI')+
  scale_y_discrete('', limits=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'), labels=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'))+scale_color_manual('', values=c('black', 'darkgrey'))


ggexport(forest_plot, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=15)
main_fig=main_fig+1
```

```{r, eval=stand_alone, fig.width=12}
forest_plot
```

# Supplemental Figure `r sup_fig`: Sensitivity analyses for controlling for different variable sets in multivariable analyses of prenatal maternal smoking and DNA methylation summary measures

```{r}
with(modeldata, table(ancestry, PostnatalMaternalSmokingAny, smkPreg_binary))%>%kbl()

forest_plot_compareModels=all_models%>%
    mutate(n=map_dbl(data, nrow), n=paste0('n=', n))%>%
    unnest(tidied)%>%
    filter(term=='smkPreg_binaryYes')%>%
    filter(outcome%in% my_outcomes)%>%
    mutate(outcome_f=factor(outcome, levels=rev(my_outcomes)))%>%
    ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=factor(age, rev(levels(age))), y=ancestry))+
    geom_point(position = position_dodge(width = 1))+
    geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
    facet_grid(predictors~outcome_f, scales='free_x', labeller = labeller(outcome_f=outcome_labels))+
    geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
    theme_classic()+
    theme(text = element_text(size=16))+theme(legend.position = 'bottom')+
    scale_x_continuous('Coefficient for prenatal maternal smoking + 95% CI')+
    scale_y_discrete('', limits=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'), labels=c('European ancestry', 'Hispanic ancestry', 'African ancestry', 'Unstratified'))+scale_color_manual('', values=c('black', 'darkgrey'))

ggexport(forest_plot_compareModels, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=15, height=15)
sup_fig=sup_fig+1
```

```{r, eval=stand_alone, fig.width=14}
forest_plot_compareModels
```

# Supplemental Figure `r sup_fig` DNA methylation summary measures across prenatal and postnatal exposures

```{r}

my_comparisons <- list(c("No prenatal smoke exposed & No maternal smoking at age 1 and 5", 
                        "No prenatal smoke exposed & Maternal smoking at age 1 or age 5"), 
                        c("No prenatal smoke exposed & Maternal smoking at age 1 or age 5", 
                          "Yes prenatal smoke exposed & Maternal smoking at age 1 or age 5"), 
                      c("Yes prenatal smoke exposed & No maternal smoking at age 1 and 5", 
                        "Yes prenatal smoke exposed & Maternal smoking at age 1 or age 5"))

n_count=modeldata%>%mutate(exposed=paste0(smkPreg_binary, ' prenatal smoke exposed & ', PostnatalMaternalSmokingAny))%>% group_by(childteen, exposed)%>%summarise(n=n())

give.n <- function(x){
  return(c(y = min(x)-0.1, label = length(x))) 
  # experiment with the multiplier to find the perfect position
}

dna_prepost=modeldata%>%
  dplyr::select(any_of(c('idnum', 'childteen', 'ancestry', 'smkPreg_binary', 'PostnatalMaternalSmokingAny', all_methyl)))%>%
  pivot_longer(cols=any_of(all_methyl),
               names_to='Methylation summary measure type',
               values_to='Methylation summary measure')%>%
  mutate(Methyl_factor=factor(`Methylation summary measure type`, 
                              levels=all_methyl_levels))%>%
  
  mutate(exposed=paste0(smkPreg_binary, ' prenatal smoke exposed & ', PostnatalMaternalSmokingAny))%>% 
  filter(Methyl_factor%in% main_methyl[3:4])%>%
  ggplot(aes(x=exposed, y=`Methylation summary measure`, color=exposed))+
    geom_violin()+geom_boxplot(width=0.2, position = position_dodge(0.9))+
  stat_summary(fun.data = give.n, geom = "text", fun.y = max,
                  position = position_dodge(width = 0.75))+
  stat_compare_means(comparisons=my_comparisons)+
    facet_grid(Methyl_factor~childteen, scales='free',labeller = labeller(Methyl_factor=all_methyl_labels))+
    scale_x_discrete(labels=function(x) str_wrap(x, width = 10))+
    theme(legend.position = 'right')+
  guides(color=guide_legend(nrow=4,byrow=TRUE))

ggexport(dna_prepost, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=10, height=10)
sup_fig=sup_fig+1
```

```{r, eval=stand_alone, fig.width=14, fig.height=10}
dna_prepost
```

# Supplemental Table `r sup_tab`: Associations between prenatal maternal smoking and DNA methylation summary measures in multivariable models among `r n_total` US children by age

```{r}
#my_main_outcomes=c('AHRR: cg05575921', 'Sustained smoking polymethylation score (newborns, cell-type controlled)', 'Sustained smoking polymethylation score (older children)', 'Pediatric clock', 'Global methylation')

test=rbind(global_models, longitudinal_global_models %>% mutate(childteen='Longitudinal'))%>%
  filter(predictors=='secondhand smoke exposure model')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(childteen=factor(case_when(childteen=='Age 9'~'Age 9 models', 
                             childteen=='Age 15'~'Age 15 models', 
                             childteen=='Longitudinal'~'Longitudinal models'),levels=c('Age 9 models', 'Age 15 models', 'Longitudinal models')), 
         tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$childteen)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_age=tbl_stack(table_list, group_header=names(table_list))


sup_tab=sup_tab+1
```

```{r, eval=stand_alone}
regression_table_age
```

# Supplemental Table `r sup_tab`: Associations between prenatal maternal smoking and DNA methylation summary measures in multivariable models among `r n_total` US children by ancestry

```{r}
test=rbind(longitudinal_global_models %>% mutate(ancestry='Unstratified'),
           longitudinal_local_models%>%dplyr::select(-data))%>%
  filter(predictors=='secondhand smoke exposure model')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$ancestry)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_ancestry=tbl_stack(table_list, group_header=names(table_list))

sup_tab=sup_tab+1

```

```{r, eval=stand_alone}
regression_table_ancestry
```

# Supplemental Figure `r sup_fig`: Surrogate variable associations with known covariates

```{r}
#bring the heatmap function
covar=c('Sample_Plate', 'Slide', 'Array', 'cm1bsex', 'smkPreg_binary', 'childteen', 'ancestry')

bring.the.heatmap.sv <- function(pd,covar,folder){
  
  nsv <- length(grep('sv',names(pd)))
  
  #generate correlation matrix between SVs and variables
  correlations <- matrix(0,nrow=length(covar),ncol=nsv)
  colnames(correlations) <- paste('sv',1:nsv,sep='')
  rownames(correlations) <- covar
  
  #take svs
  svs <- pd[,colnames(correlations)]
  
  #correlations between svs and variables:
  # cor test if continuous
  # anova is multi category
  # t-test if two category
  for(col in covar){
    var <- pd[,col]
    #check to see if it is numeric and isn't really a two category thing
    if(class(var)=='numeric' & dim(table(var))>2){
      correlations[col,] <- apply(svs,2,FUN = function(x,y) cor.test(x,y)$p.value,y=var)
    }else{
      if(length(levels(factor(var)))>2){
        correlations[col,] <- apply(svs,2,FUN = function(x,y) anova(lm(x~y))$'Pr(>F)'[1],y=var)
      }else{
        correlations[col,] <- apply(svs,2,FUN = function(x,y) t.test(x~y)$p.value,y=var)
      }
    }
  }
  color.gradient <- colorRampPalette(c('firebrick','ivory3','royalblue'),bias=3)
  correlations=correlations%>%as.data.frame()%>%tibble::rownames_to_column('var1')%>%
    pivot_longer(-var1, names_to='var2')%>%
    mutate(var2=factor(var2, levels=unique(mixedsort(var2))),
           var1=factor(var1, levels=pd[, covar]%>%colnames(), labels=get_label(pd[, covar])))
  correlations%>%
    ggplot(aes(x=var1, y=var2, fill=value))+geom_tile()+
    scale_fill_gradientn('P value', colours = color.gradient(20), values = seq(0, 1, 0.5))+
    scale_x_discrete('', labels=str_wrap(get_label(pd[, covar]), 10))+
    scale_y_discrete('')
}

pdf(file=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=11, height=13)
bring.the.heatmap.sv(modeldata, covar=c('smkPreg_binary', 'Sample_Plate', 'Slide', 'Array', 'ChildAgeComposite', 'ancestry', 'cm1bsex', 'Leukocytes_saliva', 'Epithelial.cells_saliva', 'cm1inpov', 'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'PostnatalMaternalSmokingAny', 'SmkAtVisitPastmonth'))
dev.off()
sup_fig=sup_fig+1
```

```{r, eval=stand_alone, fig.width=12, fig.height=8}
bring.the.heatmap.sv(modeldata, covar=c('smkPreg_binary', 'Sample_Plate', 'Slide', 'Array', 'ChildAgeComposite', 'ancestry', 'cm1bsex', 'Leukocytes_saliva', 'Epithelial.cells_saliva', 'cm1inpov', 'm1g2_YesNoPreg', 'm1g3_YesNoPreg', 'PostnatalMaternalSmokingAny', 'SmkAtVisitPastmonth'))
```

# Supplemental Table `r sup_tab`: Coefficients for multivariable models of prenatal maternal smoke exposure regressed on selected DNA methylation summary variables (outcomes) among children in the Fragile Families and Child Wellbeing study

```{r}
test=rbind(global_models, longitudinal_global_models %>% mutate(childteen='Longitudinal'))%>%
  filter(predictors=='surrogate variable models')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(childteen=factor(case_when(childteen=='Age 9'~'Age 9 models', 
                             childteen=='Age 15'~'Age 15 models', 
                             childteen=='Longitudinal'~'Longitudinal models'),levels=c('Age 9 models', 'Age 15 models', 'Longitudinal models')), 
         tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$childteen)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_age_svs=tbl_stack(table_list, group_header=names(table_list))


test=rbind(longitudinal_global_models %>% mutate(ancestry='Unstratified'),
           longitudinal_local_models%>%dplyr::select(-data))%>%
  filter(predictors=='surrogate variable models')%>%
  filter(outcome %in% my_outcomes)%>%
  mutate(tbl=map(model, ~tbl_regression(.x, include='smkPreg_binary', 
                                        show_single_row='smkPreg_binary', 
                                        label=list(smkPreg_binary~'Prenatal maternal smoke exposed'))%>%add_n()))

table_list=test %>%split(f=.$ancestry)%>%map(~tbl_merge(.x$tbl, 
                                                         tab_spanner = as.character(.x$outcome)))


regression_table_ancestry_svs=tbl_stack(table_list, group_header=names(table_list))

sup_tab=sup_tab+1
```

```{r, eval=stand_alone}
regression_table_ancestry_svs
```

# Figure `r main_fig` - Polymethylation scores more accurately delineate prenatal smoke exposure status than non-targeted DNA methylation summary measures and *a priori* CpGs

```{r}

age_labels=c('Age 9', 'Age 15'); names(age_labels)=c('Age 9', 'Age 15')
methyl_labels_roc=paste0('Base model+', all_methyl_labels)
names(methyl_labels_roc)=paste0('Base model+', names(all_methyl_labels))

fig4a= global_roc%>%
    dplyr::select(predictors, covariates, methylation, childteen, test, modeltype)%>%
    unnest(test) %>% 
    filter(str_detect(methylation, 'SSnewbornCT|cg0557|global') | modeltype=='Base model')%>%
    ggplot(aes(x=spec, y=sens, color=modeltype))+geom_line()+
    facet_wrap(childteen~., nrow=1, labeller=labeller(childteen=age_labels))+
    theme_classic()+
    scale_color_manual('', values=c('red', 'darkred', 'darkblue', 'darkgreen', 'blue', 'green'), 
                       function(x) str_wrap(x, width = 30))+
    theme(legend.position = 'right')+guides(color=guide_legend(nrow=4, byrow=TRUE))+
    geom_abline(intercept=1, slope=-1, color='black')+
    xlab('Specificity')+ylab('Sensitivity')+
    theme(legend.position='bottom')+ coord_fixed()

fig4b=global_roc%>%
  dplyr::select(covariates, methylation, childteen, test)%>% 
  filter(str_detect(methylation, 'SSnewbornCT') & covariates!='')%>%unnest(test)%>%
  rbind(long_global_roc%>% 
          mutate(childteen='Age 9 & Age 15', methylation=gsub('_Age_9.*', '', methylation))%>%
          dplyr::select(covariates, methylation, childteen, test)%>% 
          filter(str_detect(methylation, 'SSnewbornCT') 
                 & covariates!='')%>%
          unnest(test))%>%
  ggplot(aes(x=spec, y=sens, color=childteen, linetype=childteen))+geom_line(size=2)+
  facet_wrap(~methylation, labeller = labeller(methylation=c('SSnewbornCT_center'='Polymethylation score, \n newborn mean centered')))+
  theme_classic()+theme(legend.position = 'right')+guides(color=guide_legend(nrow=4,byrow=TRUE))+
  geom_abline(intercept=1, slope=-1, color='black')+
  scale_color_grey('', labels=c('Age 9 only', 'Age 9 & 15', 'Age 15 only'))+
  scale_linetype('', labels=c('Age 9 only', 'Age 9 & 15', 'Age 15 only'))+
  xlab('Specificity')+ylab('Sensitivity')+
  theme(legend.position='bottom')+ coord_fixed()

auc_fig=cowplot::plot_grid(fig4a, fig4b, labels=c('A', 'B'), nrow=1, rel_widths=c(0.65, 0.35), align='h', axis='t')

ggexport(auc_fig, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')), width=9)
main_fig=main_fig+1
```

```{r, eval=stand_alone, fig.width=10}
auc_fig
```

# Supplemental Table with AUC and p-value comparison to sustained smoking, newborn cell type controls polymethylation score

```{r}
roc_table=global_roc %>% dplyr::select(childteen, methylation, modeltype, roc, auc)
roc_base=roc_table %>% filter(modeltype=='Base model+No methylation')%>%mutate(roc_basemodel=roc)%>%dplyr::select(childteen, roc_basemodel)
roc_ss=roc_table %>% filter(modeltype=='Sustained smoking w/ cell type correction (newborn)')%>%mutate(roc_pms=roc)%>%dplyr::select(childteen, roc_pms)

roc_table1=roc_table%>%filter(!str_detect(modeltype, 'Base model'))%>%left_join(roc_ss)%>%mutate(roctest=map2(roc, roc_pms, ~roc.test(.x, .y)))%>%mutate(auc=round(auc, 2))

rct1=roc_table1%>%mutate(`pval comparison to sustained smoking score (newborns)`=map_dbl(roctest, ~round(.x$p.value, 4)))%>%dplyr::select(childteen, modeltype, auc, `pval comparison to sustained smoking score (newborns)`)

```

```{r, eval=stand_alone}
rct1%>%kbl(col.names=c('Age', 'Model', 'AUC', 'P value (vs polymethylation score, newborns, cell type controlled)'))
```


# Supplemental Table with AUC and p-value comparison to base model with no methylation variables (i.e., using only child sex, income to poverty ratio, saliva/epithelial cell proportions and sample plate)

```{r}

roc_table2=roc_table%>%filter(str_detect(modeltype, 'Base model'))%>%left_join(roc_base)%>%mutate(roctest=map2(roc, roc_basemodel, ~roc.test(.x, .y)))%>%mutate(auc=round(auc, 2))

rct2=roc_table2%>%mutate(`pval comparison to base model`=map_dbl(roctest, ~round(.x$p.value, 4)))%>%dplyr::select(childteen, modeltype, auc, `pval comparison to base model`)

rct2%>%kbl(col.names=c('Age', 'Model', 'AUC', 'P value (vs base w/ no methyl)'))


```

```{r, eval=stand_alone}
rct2%>%kbl(col.names=c('Age', 'Model', 'AUC', 'P value (vs base w/ no methyl)'))
```

# Question for group: add EWAS results to this paper

-   Arguments in favor: recent paper in Clinical Epigenetics included both
-   Arguments against: enough here for a compelling story; deadline of December 1st; submit and if asked for we have it available.

```{r, eval=include_ewas}

calcLambda<-function(df, type='lm'){
  if(type=='lme'){df=df%>%mutate(P.Value=p.value)}
  o<- -log10(sort(df$P.Value, decreasing = F))
  e<- -log10(ppoints(length(df$P.Value)))
  lambda<-median(o)/median(e)
  return(lambda)
}
```

```{r, eval=include_ewas}
child_m1=readRDS(file.path(datadir, 'CreatedData', 'EWAS', 'child1_TopHits.RDS'))
teen_m1=readRDS(file.path(datadir, 'CreatedData', 'EWAS', 'teen1_TopHits.RDS'))
```

```{r, include=include_ewas}
#lme_base=readRDS(file.path(datadir, 'CreatedData', 'EWAS', 'lme_ewas_base.RDS'))
#lme_smk = lme_base %>% filter(term=='smkPreg_binaryYes')%>%mutate(p.bh=p.adjust(p.value, 'hochberg'))
#lme_smk=add_annotations(lme_smk, type='lme')
lme_smk=readRDS(file.path(datadir, 'CreatedData', 'EWAS', 'lme_smoking.RDS'))
#mutate columns to get matching columns in lme_smk as in child & teen _m1
lme_smk = lme_smk %>% mutate(logFC=estimate, CI.L=conf.low, CI.R=conf.high, P.Value=p.value, adj.P.Val=p.bh)
```

```{r, eval=F}
library(variancePartition)
dream=readRDS(file.path(datadir, 'CreatedData', 'EWAS', 'dream_ewas.RDS'))
dream_smk=topTable(dream, coef='smkPreg_binaryYes', number=nrow(dream$p.value))
#dream_smk=add_annotations(dream_smk)
```

## QQ plots

```{r, eval=include_ewas}

png(file=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.png')))
par(mfrow=c(3, 1), mar=c(5.1, 4.1, 5.1, 2.1))
  qqman::qq(child_m1$P.Value, main='Age 9: Base model EWAS')
  mtext(paste0(expression(lambda), '=', round(calcLambda(child_m1), 3)), side=3)
  
  qqman::qq(teen_m1$P.Value, main='Age 15: Base model EWAS')
  mtext(paste0(expression(lambda), '=', round(calcLambda(teen_m1), 3)), side=3)
  
  qqman::qq(lme_smk$P.Value, main='Mixed effect models: Base model EWAS')
  mtext(paste0(expression(lambda), '=', round(calcLambda(lme_smk), 3)), side=3)
dev.off()

```

```{r, eval=stand_alone, eval=include_ewas}
par(mfrow=c(3, 1), mar=c(5.1, 4.1, 5.1, 2.1))
  qqman::qq(child_m1$P.Value, main='Age 9: Base model EWAS')
  mtext(paste0(expression(lambda), '=', round(calcLambda(child_m1), 3)), side=3)
  
  qqman::qq(teen_m1$P.Value, main='Age 15: Base model EWAS')
  mtext(paste0(expression(lambda), '=', round(calcLambda(teen_m1), 3)), side=3)
  
  qqman::qq(lme_smk$P.Value, main='Mixed effect models: Base model EWAS')
  mtext(paste0(expression(lambda), '=', round(calcLambda(lme_smk), 3)), side=3)

sup_fig=sup_fig+1
```



## Beta vs beta plot: Age 9 vs age 15

```{r, cache=T, eval=include_ewas}
library(tibble)
base_models=list('Age_9'=child_m1%>%rownames_to_column(var='CPG'), 'Age_15'=teen_m1%>%rownames_to_column(var='CPG'), "Mixed_effect_model"=lme_smk)%>%Map(cbind, ., ModelType=names(.))

beta_base<-lapply(base_models, function(x) x %>%dplyr::select(CPG, logFC, ModelType)%>%pivot_wider(id_cols = CPG, values_from=logFC, names_from=ModelType))%>%purrr::reduce(left_join, by='CPG')
  

smooth_betas<-function(df, x, y){
  ggplot(df, aes_string(x=x, y=y))+
    geom_point()+geom_smooth(method='lm')+
    stat_cor()+theme_classic()+
    scale_y_continuous(name=y)+
   scale_x_continuous(name=x)+
    coord_fixed()
}

sig_beta_m1<-child_m1%>%rownames_to_column(var='CPG')%>%filter(P.Value<10e-4)%>%pull(CPG)

beta_compare=ggarrange( 
  smooth_betas(beta_base, "Age_9", "Age_15"),
  smooth_betas(beta_base %>%filter(CPG %in% sig_beta_m1), 'Age_9', 'Age_15'), 
  labels=c('A', 'B'), align='h')

beta_compare
ggexport(beta_compare, filename = file.path(outdir, paste0('SupplementalFigure', sup_fig, '.png'))) 
```

```{r, eval=stand_alone, eval=include_ewas}
beta_compare
```

## Upset plots

```{r, eval=include_ewas}
library(ggupset)
sig_cpg_list<-map(base_models, ~.x %>%filter(P.Value<10e-4)%>%dplyr::select(CPG, ModelType))%>%
  bind_rows()%>%
  group_by(CPG)%>%summarise(Models=list(ModelType))
  
upset_cpgs=ggplot(sig_cpg_list, aes(x=Models))+
    geom_bar() +
    scale_x_upset()
ggexport(upset_cpgs, filename = file.path(outdir, paste0('SupplementaalFigure', sup_fig, '.pdf')))
sup_fig=sup_fig+1
```

## Volcano plot

```{r, eval=include_ewas}
volcs=ggarrange(
ggplot(child_m1%>%mutate(adjPval=ifelse(adj.P.Val<0.05, '<0.05', ifelse(adj.P.Val<0.2, '<0.2', '>0.2'))), aes(x=logFC, y=-log10(P.Value), color=adjPval))+geom_point()+scale_x_continuous('Beta (regression coefficient)')+
  scale_color_manual('Benjamini-Hochberg adjusted P value', values=c('red', 'orange', 'black'))+
  scale_y_continuous(limits=c(0, 9), breaks=c(0, 2, 4, 6, 8))+
  scale_x_continuous(limits=c(-0.06, 0.06), breaks=c(-0.05, 0, 0.05)),
ggplot(teen_m1%>%mutate(adjPval=ifelse(adj.P.Val<0.05, '<0.05', ifelse(adj.P.Val<0.2, '<0.2', '>0.2'))), aes(x=logFC, y=-log10(P.Value), color=adjPval))+geom_point()+scale_x_continuous('Beta (regression coefficient)')+
  scale_color_manual('Benjamini-Hochberg adjusted P value', values=c('red', 'orange', 'black'))+
  scale_y_continuous(limits=c(0, 9), breaks=c(0, 2, 4, 6, 8))+
  scale_x_continuous(limits=c(-0.06, 0.06), breaks=c(-0.05, 0, 0.05)),
ggplot(lme_smk%>%mutate(adjPval=ifelse(adj.P.Val<0.05, '<0.05', ifelse(adj.P.Val<0.2, '<0.2', '>0.2'))), aes(x=logFC, y=-log10(P.Value), color=adjPval))+geom_point()+scale_x_continuous('Beta (regression coefficient)')+
  scale_color_manual('Benjamini-Hochberg adjusted P value', values=c('red', 'orange', 'black'))+
  scale_y_continuous(limits=c(0, 9), breaks=c(0, 2, 4, 6, 8))+
  scale_x_continuous(limits=c(-0.06, 0.06), breaks=c(-0.05, 0, 0.05)),
labels=c('A', 'B', 'C'), 
common.legend = TRUE, 
nrow=1
)

ggexport(volcs, filename=file.path(outdir, paste0('Figure', main_fig, '.pdf')))
ggexport(volcs, filename=file.path(outdir, paste0('Figure', main_fig, '.png')))
```

```{r, eval=stand_alone, eval=include_ewas, fig.width=14, }
volcs
```

## Gap hunter

```{r, eval=include_ewas}
all_gaps=readRDS(file.path(datadir, 'CreatedData', 'EWAS', 'gaphunter_gaps.RDS'))
```

## Compare across models and studies

```{r, eval=include_ewas}
topCPGs<-read.csv(paste0(datadir, "/TopCpGsWiklundEtAl2019.csv"))
topCPGs<-topCPGs %>% mutate(estimate=as.numeric(gsub("−\\s", "-", gsub(" .*", "", β..SE.))), std.error=as.numeric(sub("\\).*", "", sub(".*\\(", "", β..SE.))), p.value=P.value, source="Wiklund et al. 2019", conf.low=(estimate-1.96*std.error), conf.high=(estimate+1.96*std.error), set ='')%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)

#read in Joubert meta top cpgs 
#read in cpgs from PMC4833289
cpgs<-read.csv(paste0(datadir, '/PMC4833289_SuppFile2CpGs.csv'), stringsAsFactors = F, header=F, skip=2)
#make headers
header1<-scan(paste0(datadir, '/PMC4833289_SuppFile2CpGs.csv'), nlines=1, what=character(), sep=',')
header1[7:26]<-c(rep("SSnewborn", 5), rep("SSnewbornCT", 5), rep("SSolder", 5), rep("anynewborn", 5))
header2<-paste(header1, scan(paste0(datadir, '/PMC4833289_SuppFile2CpGs.csv'), nlines=1, skip=1, what=character(), sep=','), sep='_')
names(cpgs)<-header2

#four meta-EWASes: SS_newborn (sustained smoking in newborns); SS_newbornCT (sustained smoking in newborns controlling for cell type) SS_older (sustained smoking in older children); any_newborn (any smoking in newborns)
#pivot_longer to get single column of coefs and a column for the analysis type
#and group into a list by analysis type
cpgs<-cpgs %>% pivot_longer(cols = c(matches("_Coef|_SE|_P|_FDR|_Direction")), names_to = c("set", ".value"), names_pattern="(.+)_(.+)")%>%rename_with(~gsub('_', '', .x))%>%mutate(Gene=`Mapped Gene`, estimate=Coef, conf.low=estimate-1.96*SE, conf.high=estimate+1.96*SE, source='Joubert et al. 2016', std.error=SE, p.value=P, Chr=CHR) %>%mutate(source=paste0(source, ': ', set))%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)%>%filter(set %in% c('SSolder', 'SSnewbornCT'))


#myCPGs<-AllModelsdf %>%filter(Model=='model1')%>%mutate(data=map(data, function(x) x %>%filter(CpG %in% sigCpGs)))%>%unnest(cols=c(data))%>%mutate(Joubertetal=ifelse(CpG %in% cpgs$CpG, 'in Joubert et al. 2016', 'Not in Joubert et al. 2016'), Wicklundetal=ifelse(CpG %in% topCPGs$CpG, 'in Wicklund et al. 2019', 'Not in Wicklund et al. 2019'))%>%mutate(Gene=gene, estimate=logFC, conf.low=CI.L, conf.high=CI.R, source=paste0('Fragile Families:', Age), std.error='NA', p.value=P.Value, Chr=chr, set='Model1', Position=pos)


```

```{r, fig.width=14, eval=include_ewas}
comparison=left_join(
  lme_smk%>%dplyr::select(CPG, gene, estimate, p.value)%>%dplyr::rename_with(~paste0('FF_', .x), .cols=c(estimate, p.value)), 
  cpgs%>%filter(set=='SSnewbornCT')%>%mutate(CPG=CpG)%>%dplyr::select(CPG, Gene, estimate, p.value)%>%dplyr::rename_with( ~paste0('Joubert_', .x), .cols=c(estimate, p.value))
  )%>%filter(!is.na(Joubert_estimate))

comp_nofilter=ggplot(comparison, aes(x=Joubert_estimate, y=FF_estimate, color=FF_p.value))+
  geom_point()+stat_cor()+
  scale_y_continuous('Estimate in Fragile Families')+
  scale_x_continuous('Estimate in Joubert et al. metaanalysis')+
  scale_color_continuous('P value in Fragile Families')

#comparison %>% mutate(s1=sign(FF_estimate), s2=sign(Joubert_estimate), agreement=case_when(s1==s2~'Direction consistent', s1<s2~'FF -, Joubert +', s1>s2~'FF+, Joubert-'))%>%group_by(agreement)%>%summarise(n=n())

comp_filter=ggplot(comparison%>%filter(FF_p.value<10e-4), aes(x=Joubert_estimate, y=FF_estimate, color=FF_p.value, label=ifelse(abs(Joubert_estimate)>0.04 , gene, NA)))+
  geom_point()+stat_cor()+
  geom_text_repel()+
  scale_y_continuous('Estimate in Fragile Families')+
  scale_x_continuous('Estimate in Joubert et al. metaanalysis')+
  scale_color_continuous('P value in Fragile Families')


comp_joubert=ggarrange(comp_nofilter, comp_filter, common.legend = F, labels=c('A', 'B'))
ggexport(comp_joubert, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.png')))
```

```{r, eval=stand_alone, eval=include_ewas, fig.width=14}
comp_joubert
```

## Additional covariate control and ancestry stratified models at top hits

```{r, eval=include_ewas}
top_cpgs = map(base_models, ~.x %>%filter(adj.P.Val<0.05)%>%dplyr::select(CPG, ModelType, gene))%>%bind_rows()%>%mutate(label=paste0(gene, ': ', CPG))
top_cpg_gaps=top_cpgs[top_cpgs$CPG %in% rownames(all_gaps$proberesults),]
top_cpgs=top_cpgs[!(top_cpgs$CPG %in% rownames(all_gaps$proberesults)),]
top_cpg_groups=top_cpgs %>%group_by(CPG)%>%dplyr::summarise(models=list(ModelType))
top_cpg_names=c(top_cpgs%>%pull(label)%>%unique(), 'AHRR: cg05575921', 'MYO1G: cg22132788')
top_cpgs=c(top_cpgs %>% pull(CPG)%>%unique(), 'cg05575921', 'cg22132788')

betaqc<-readRDS(file=file.path(datadir, 'OGData', "betaqc.rds"))
my_beta=as.data.frame(t(betaqc[row.names(betaqc)%in% top_cpgs, ])) %>% mutate(MethID=row.names(.))
modeldata=left_join(modeldata, my_beta)
```

```{r, eval=F}
y_vector<-top_cpgs
outcome_labels=top_cpg_names
names(outcome_labels)=y_vector

#model variables (excepting PCs, add in individually for global & ancestry specific models)
base_model_vars<-"~smkPreg_binary+cm1bsex+cm1inpov+Leukocytes_saliva+Epithelial.cells_saliva+Sample_Plate"
prenatal_model_vars<-paste0(base_model_vars, "+m1g2_YesNoPreg+m1g3_YesNoPreg")
secondhand_model_vars<-paste0(prenatal_model_vars, '+PostnatalMaternalSmokingAny+SmkAtVisitPastmonth')
interaction_model_vars<-paste0(secondhand_model_vars, '+ChildAgeComposite:smkPreg_binary')
#pc variables
global_pcs='+global_PC1+global_PC2'
local_pcs='+local_PC1+local_PC2'

#surrogate variable models
surrogate_model_vars<-paste0('~smkPreg_binary+', 
                             paste(colnames(modeldata)[str_detect(colnames(modeldata), 'sv')], collapse='+'))

predictors=c('base model'=base_model_vars, 
             'prenatal exposures model'=prenatal_model_vars, 
             'secondhand smoke exposure model'=secondhand_model_vars)

predictors_interaction=c(predictors, 'interaction model'=interaction_model_vars)

model_labels=c('base model', 'prenatal exposures model', 'secondhand smoke exposure model', 'surrogate variable models')

model_labels_interaction=c('base model', 'prenatal exposures model', 'secondhand smoke exposure model', 'interaction model (age*smoke)', 'surrogate variable models')

#stratification variables 
##############################
age_labels=c('Age 9', 'Age 15')
age_vector=modeldata$childteen%>%unique()
names(age_labels)=age_vector
ancestry_vector=modeldata$ancestry%>%unique()

#################################Source functions#######################################
model<-function(df, y, x){
  lm(formula(paste(y, x)), data=df)
}

model_lmm<-function(df, y, x){
  xnew=paste0(x, '+', '(1|id)')
  lmer(formula(paste(y, xnew)), data=df, REML=F)
}

my_clean_argnames=function(df, my_preds, childteen=T, model_labs=model_labels){
  names(model_labs)=my_preds
  df=df %>% mutate(predictors=recode_factor(predictors, !!!model_labs),
                outcome=recode_factor(outcome, !!!outcome_labels))
  if(childteen==T){df=df%>%mutate(age=recode_factor(childteen, !!!age_labels))} 
  return(df)
}
  
#################################Cross sectional models###########################
#########################Global
#set predictor variables & arguments
global_predictors=c(paste0(predictors, global_pcs), surrogate_model_vars)
args=list('childteen'=age_vector, 'outcome'=top_cpgs, 'predictors'=global_predictors)%>%cross_df()

#run models
global_models=modeldata%>%group_by(childteen)%>%nest()%>%left_join(args)%>%
  mutate(model=pmap(list(data, outcome, predictors), model), 
         tidied=map(model, tidy, conf.int=T), 
         glanced=map(model, glance))

#clean argument names for plotting
global_models=my_clean_argnames(global_models, global_predictors)

#########################Local
#arguments
local_predictors=c(paste0(predictors, local_pcs), surrogate_model_vars)
args=list('childteen'=c('Age 9', 'Age 15'), 'outcome'=top_cpgs, 'predictors'=local_predictors, 'ancestry'=ancestry_vector)%>%cross_df()

#run models
local_models=modeldata%>%group_by(childteen, ancestry)%>%nest()%>%left_join(args)%>%
  mutate(model=pmap(list(data, outcome, predictors), model), 
         tidied=map(model, tidy, conf.int=T), 
         glanced=map(model, glance))

#clean argument names for plotting
local_models=my_clean_argnames(local_models, local_predictors)

#################################Longitudinal models##################
#########################Global
#set predictor variables & arguments
#include a model with an interaction term b/t age and prenatal smoke exposure
#to determine if there is a age-slope difference b/t exposed & not exposed
global_predictors=c(paste0(predictors_interaction, global_pcs, '+ChildAgeComposite'), surrogate_model_vars)
args=list('outcome'=top_cpgs, 'predictors'=global_predictors)%>%cross_df()

#run models
longitudinal_global_models=args %>% 
    mutate(model=pmap(list(y=outcome, x=predictors), model_lmm, df=modeldata), 
           tidied=map(model, tidy, conf.int=T), 
           glanced=map(model, glance))

#clean argument names
longitudinal_global_models=my_clean_argnames(longitudinal_global_models, global_predictors, childteen = F, model_labs = model_labels_interaction)
#########################Local
#set predictor variables and arguments
local_predictors=c(paste0(predictors_interaction, local_pcs, '+ChildAgeComposite'), surrogate_model_vars)
args=list('outcome'=top_cpgs, 'predictors'=local_predictors, 'ancestry'=ancestry_vector)%>%cross_df()

#run models
longitudinal_local_models=modeldata%>%group_by(ancestry)%>%nest()%>%left_join(args)%>%
  mutate(model=pmap(list(data, outcome, predictors), model_lmm), 
         tidied=map(model, tidy, conf.int=T), 
         glanced=map(model, glance))

#clean argument names
longitudinal_local_models=my_clean_argnames(longitudinal_local_models, local_predictors, childteen = F, model_labs = model_labels_interaction)
```

```{r compare over time, eval=F}

rbind(global_models%>%dplyr::select(childteen, predictors, outcome, tidied)%>%unnest(tidied)%>%filter(term=='smkPreg_binaryYes'), 
      longitudinal_global_models%>%dplyr::select(predictors, outcome, tidied)%>%unnest(tidied)%>%filter(term=='smkPreg_binaryYes')%>%mutate(childteen='Linear mixed model'))%>%
  filter(predictors %in% c('base model', 'secondhand smoke exposure model'))%>%
  ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=childteen, y=predictors))+
    geom_point(position = position_dodge(width = 1))+
    geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
    facet_grid(.~outcome, scales='free_x', labeller = labeller(outcome_f=top_cpg_names))+
    geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
    theme_classic()


cpg_timevariant=top_cpg_groups%>%mutate(count=map_dbl(models, length))%>%filter(count<3)%>%pull(CPG)

age_long_methyl%>%
  filter(`Methylation summary measure type` %in% top_cpgs)%>%
  ggplot(aes(x=childteen, y=`Methylation summary measure`, color=childteen))+
  geom_violin()+geom_boxplot(width=0.2)+stat_compare_means(label="..p.signif..")+
  scale_color_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  scale_x_discrete(name='Visit', labels=c('Age 9', 'Age 15'))+
  geom_point(alpha=0)+geom_line(alpha=0, aes(group=idnum))+geom_smooth(aes(group=1), method='lm', color='black')+
  facet_wrap(~Methyl_factor, scales='free')+
  theme(legend.position = 'bottom')


```

```{r compare over ancestries, eval=F}
longitudinal_local_models%>%unnest(tidied)%>%filter(term=='smkPreg_binaryYes')%>%
  filter(predictors %in% c('base model', 'secondhand smoke exposure model'))%>%
  ggplot(aes(x=estimate, xmin=conf.low, xmax=conf.high, color=ancestry, y=predictors))+
    geom_point(position = position_dodge(width = 1))+
    geom_errorbarh(height=0.1, position = position_dodge(width = 1))+
    facet_grid(.~outcome, scales='free_x', labeller = labeller(outcome_f=top_cpg_names))+
    geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+
    theme_classic()


longitudinal_local_models%>%unnest(tidied)%>%filter(term=='smkPreg_binaryYes')%>%
    filter(predictors %in% c('base model'))%>%group_by(outcome)%>%summarise(est=list(estimate))%>%mutate(sign=map_dbl(est, ~sum(sign(.x))))%>%filter(abs(sign)!=3)

```

```{r, fig.width=14, eval=include_ewas}

age_long_methyl=modeldata%>%
  dplyr::select(any_of(c('idnum', 'childteen', 'ancestry', 'smkPreg_binary', top_cpgs)))%>%
  pivot_longer(cols=any_of(top_cpgs),
               names_to='Methylation summary measure type',
               values_to='Methylation summary measure')%>%
  mutate(Methyl_factor=factor(`Methylation summary measure type`, 
                              levels=top_cpgs))

ancestry_differences=age_long_methyl%>%
    filter(`Methylation summary measure type` %in% top_cpgs)%>%
    ggplot(aes(x=ancestry, y=`Methylation summary measure`, color=smkPreg_binary))+
    geom_violin()+geom_boxplot(width=0.2, position=position_dodge(0.9))+stat_compare_means(label="..p.signif..", label.y.npc = 0.95)+
    scale_color_discrete(name='')+
    scale_x_discrete()+
    geom_point(alpha=0)+
    facet_grid(childteen~Methyl_factor, scales='free')+theme_bw()+
    theme(legend.position = 'bottom')+
  theme(axis.text.x = element_text(angle=45, hjust=1))

ancestry_differences

ggexport(ancestry_differences, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')), width=11)

ggexport(ancestry_differences, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.png')))
```

```{r, eval=stand_alone, eval=include_ewas, fig.width=14}
ancestry_differences
```

```{r, eval=include_ewas}

myCPGs<-base_models%>%bind_rows(.id='modeltype')%>%mutate(CpG=CPG, Gene=gene, estimate=logFC, conf.low=CI.L, conf.high=CI.R, source=paste0('Fragile Families:', modeltype), std.error='NA', p.value=P.Value, Chr=chr, set='Model1', Position=pos)%>%dplyr::select(colnames(cpgs))

#cpg_overlap=myCPGs%>%filter(Joubertetal=='in Joubert et al. 2016' | Wicklundetal=='in Wicklund et al. 2019')%>%pull(CpG)%>%unique()

#myCPGs2<-myCPGs %>%ungroup()%>%dplyr::select(CpG, Chr, Position, Gene, estimate, std.error, p.value, conf.low, conf.high, source, set)%>%mutate(CpG=ifelse(CpG %in% row.names(mygaps$proberesults), paste0(CpG, '*'), CpG))

cpg_all<-rbind(myCPGs%>%filter(CpG %in% top_cpgs), cpgs%>%filter(CpG %in% top_cpgs), topCPGs%>%filter(CpG%in%top_cpgs))


#geom blank to give good axes limits
blank_data <- cpg_all %>%dplyr::select(Gene, CpG, estimate, std.error)%>%group_by(Gene)%>%mutate(min=min(estimate), max=max(as.numeric(std.error), na.rm=T))%>%dplyr::select(-estimate, -std.error)%>%unique()%>%mutate(estimate=min*-1+max, conf.high=estimate, conf.low=estimate)%>%dplyr::select(-max, -min)%>%arrange(Gene)%>%mutate(source='Wiklund et al. 2019')


blank_data2 <- cpg_all %>%dplyr::select(Gene, CpG, estimate, std.error)%>%group_by(Gene)%>%mutate(min=min(estimate))%>%dplyr::select(-estimate, -std.error)%>%unique()%>%mutate(estimate=min*-1+0.005, conf.high=estimate, conf.low=estimate)%>%dplyr::select(-min)%>%arrange(Gene)%>%mutate(source='Fragile Families: Age_9')

plot_sharedCpG<-ggplot(cpg_all, aes(x=estimate, xmin=conf.low, xmax=conf.high, color=source, y=CpG))+geom_point(position=position_dodge(width=1))+geom_errorbarh(height=0.1, position=position_dodge(width=1))+scale_x_continuous()+geom_vline(xintercept=0, color='red', linetype="dashed", alpha=0.5)+facet_wrap(.~Gene, scales='free', ncol=2)+theme_classic()+theme(legend.position = 'bottom')+guides(color=guide_legend(nrow=4,byrow=TRUE))+geom_blank(data = blank_data, aes(x = estimate))+scale_color_discrete('', labels=c('Fragile families saliva: age 15', 'Fragile families saliva: age 9', 'Fragile families: mixed effect model', 'Joubert et al. 2016 newborn cord blood', 'Joubert et al. 2016 peripheral blood older children', 'Wicklund et al. peripheral blood >16 years'))


ggexport(plot_sharedCpG, filename=file.path(outdir, paste0('SupplementalFigure', sup_fig, '.pdf')))
```

```{r, eval=stand_alone, eval=include_ewas, fig.width=14}
plot_sharedCpG
```
